<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>CosMx Data Loading and Seurat Object Creation</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #204a87; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #204a87; font-weight: bold; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #ce5c00; font-weight: bold; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>







<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">CosMx™ SMI Data Analysis Pipeline</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="dataset.html">Dataset</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pipeline
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="pipeline.html">Introduction</a>
    </li>
    <li>
      <a href="0.0_data_loading.html">Data Loading</a>
    </li>
    <li>
      <a href="1.0_qc_and_filtering.html">QC and filtering</a>
    </li>
    <li>
      <a href="2.0_normalization.html">Normalization</a>
    </li>
    <li>
      <a href="3.0_dimensional_reduction.html">Dimensional reduction</a>
    </li>
    <li>
      <a href="4.0_insitutype_cell_typing.html">Cell typing</a>
    </li>
  </ul>
</li>
<li>
  <a href="norm_examples.html">Normalization</a>
</li>
<li>
  <a href="clus_examples.html">Cell typing</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
<li>
  <a href="https://github.com/lidiaga/CosMx_pipeline_LGA">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">CosMx Data Loading and Seurat Object
Creation</h1>
<h4 class="author">LGA</h4>
<h4 class="date">02 septiembre 2025</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-09-02
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>CosMx_pipeline_LGA/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20250517code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20250517)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20250517code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20250517)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrong9bc387c">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong> 9bc387c
</a>
</p>
</div>
<div id="strongRepositoryversionstrong9bc387c"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version 9bc387c.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    NBClust-Plots/
    Ignored:    analysis/.Rhistory
    Ignored:    data/seuAtoMx/
    Ignored:    output/processed_data/Log/
    Ignored:    output/processed_data/RC/
    Ignored:    output/processed_data/SCT/
    Ignored:    output/processed_data/exprMat_unfiltered.RDS
    Ignored:    output/processed_data/fov_positions_unfiltered.RDS
    Ignored:    output/processed_data/metadata_unfiltered.RDS
    Ignored:    output/processed_data/negMat_unfiltered.RDS
    Ignored:    output/processed_data/seu_filtered.RDS
    Ignored:    output/processed_data/seu_semifiltered.RDS

Untracked files:
    Untracked:  output/performance_reports/

Unstaged changes:
    Modified:   CosMx_pipeline_LGA.Rproj
    Modified:   _workflowr.yml

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/0.0_data_loading.Rmd</code>) and
HTML (<code>docs/0.0_data_loading.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
78c8449
</td>
<td>
lga-uoc
</td>
<td>
2025-09-02
</td>
<td>
Add Rmds files in analysis
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<!-- This pipeline has been developed using Seurat v4.4, as Seurat v5 presents several changes to the object structure that affected the working methods. Additionally, the AtoMx exported Seurat objects come in the Seurat v4 structure. Therefore, working in this version can be more compatible with exported data too. -->
<div id="dependencies" class="section level1">
<h1>Dependencies</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(data.table) <span class="co"># Efficient data management</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(Matrix) <span class="co"># Sparse matrices</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">library</span>(here) <span class="co"># Enhanced file referencing in project-oriented workflows</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">library</span>(vecsets) <span class="co"># Enhanced version of the Set Operations from &quot;base&quot;</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="fu">library</span>(dplyr) <span class="co"># For the use of pipes %&gt;%</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="fu">library</span>(kableExtra) <span class="co"># For table formatting</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="fu">library</span>(Seurat) <span class="co"># Seurat object</span></span></code></pre></div>
</div>
<div id="load-the-data" class="section level1">
<h1>Load the data</h1>
<p>First of all, flat files directory needs to be located in the folder
structure of our project.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="do">## Code adapted from Scratch Space vignette</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="co"># Locate flat files</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>flatFiles_dir <span class="ot">&lt;-</span> <span class="fu">here</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;flatFiles&quot;</span>)</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>slideNames <span class="ot">&lt;-</span> <span class="fu">dir</span>(flatFiles_dir)</span></code></pre></div>
<ul>
<li><strong>Total number of slides in experiment:</strong> 1.<br />
</li>
<li><strong>Slide Names:</strong> CoronalHemisphere.</li>
</ul>
<p>In this case, there is only one slide, this will simplify early
pipeline construction. Future improvements could include: adapt the code
to work with multiple slides and perform integration.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="do">## Code adapted from Scratch Space vignette</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="co"># Locate slide files</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>slide <span class="ot">&lt;-</span> slideNames[<span class="dv">1</span>]</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>slide_dir <span class="ot">&lt;-</span> <span class="fu">here</span>(flatFiles_dir, slide)</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>slide_files <span class="ot">&lt;-</span> <span class="fu">dir</span>(slide_dir, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="co"># Load slide files</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="do">### Fov_positions</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>fov_positions <span class="ot">&lt;-</span> <span class="fu">fread</span>(slide_files[<span class="fu">grep</span>(<span class="st">&quot;fov_positions_file.csv&quot;</span>, slide_files)], <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="do">## fread() is an efficient version of read.csv, that is able to read gzip files</span></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a><span class="do">## and creates a data.table object for efficient management</span></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a><span class="do">### Metadata</span></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">fread</span>(slide_files[<span class="fu">grep</span>(<span class="st">&quot;metadata_file.csv&quot;</span>, slide_files)], <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a></span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a><span class="do">### Counts matrix (efficient loading)</span></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a>exprMat_dir <span class="ot">&lt;-</span> slide_files[<span class="fu">grep</span>(<span class="st">&quot;exprMat_file.csv&quot;</span>, slide_files)]</span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a>exprMat <span class="ot">&lt;-</span> <span class="fu">loadExprMat</span>(exprMat_dir, <span class="at">slide_number =</span> <span class="dv">1</span>)</span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a></span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a><span class="do">## loadExprMat() is a self created function that reads the expression matrix</span></span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a><span class="do">## in chunks and converts it to sparse (adapted from Scratch Space vignette)</span></span>
<span id="cb3-25"><a href="#cb3-25" tabindex="-1"></a><span class="do">## available code is in &quot;code/aux_functions.R&quot;</span></span></code></pre></div>
</div>
<div id="explore-the-data" class="section level1">
<h1>Explore the data</h1>
<div id="fov-positions" class="section level2">
<h2>FOV-positions</h2>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Show head FOV positions</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">head</span>(fov_positions)</span></code></pre></div>
<pre><code>     fov x_global_px y_global_px
   &lt;int&gt;       &lt;num&gt;       &lt;num&gt;
1:     1   -498383.3    6475.000
2:     2   -498383.3    2208.333
3:     3   -498383.3   -2058.333
4:     4   -498383.3   -6325.000
5:     5   -498383.3  -10591.667
6:     6   -498383.3  -14858.333</code></pre>
<ul>
<li><strong>Dimensions:</strong> 130 FOVs and 3 columns.<br />
</li>
<li><strong>FOVs:</strong> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,
14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31,
32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49,
50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85,
86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102,
103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116,
117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129,
130.<br />
</li>
<li><strong>Columns:</strong> fov, x_global_px, y_global_px.</li>
</ul>
<p>At first sight, there are no signs of previous pre-processing or
missing FOVs.</p>
</div>
<div id="metadata" class="section level2">
<h2>Metadata</h2>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Show head metadata</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>metadata[<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]</span></code></pre></div>
<pre><code>     fov cell_ID  Area AspectRatio CenterX_local_px CenterY_local_px
   &lt;int&gt;   &lt;int&gt; &lt;int&gt;       &lt;num&gt;            &lt;int&gt;            &lt;int&gt;
1:     1       1  6073        0.47             4222             4066
2:     1       2  5675        0.72             4182             3938
3:     1       3 12896        1.26             2156             3864
4:     1       4  8234        0.51             4108             3608
5:     1       5  9852        0.88             4162             3506
6:     1       6 13372        0.90             4167             3301</code></pre>
<ul>
<li><strong>Dimensions:</strong> 48556 cells and 20 attributes.<br />
</li>
<li><strong>FOVs:</strong> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,
14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31,
32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49,
50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85,
86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102,
103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116,
117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129,
130.<br />
</li>
<li><strong>Columns:</strong> fov, cell_ID, Area, AspectRatio,
CenterX_local_px, CenterY_local_px, CenterX_global_px,
CenterY_global_px, Width, Height, Mean.Histone, Max.Histone, Mean.G,
Max.G, Mean.rRNA, Max.rRNA, Mean.GFAP, Max.GFAP, Mean.DAPI,
Max.DAPI.</li>
</ul>
<p>In sight of the file columns, there are no signs of previous
pre-processing or missing FOVs.</p>
</div>
<div id="expression-matrix" class="section level2">
<h2>Expression Matrix</h2>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># Check presence of rows with cell_id of 0</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>cells_0 <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">&quot;_0$&quot;</span>, <span class="fu">rownames</span>(exprMat), <span class="at">value =</span> <span class="cn">TRUE</span>) </span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="co"># Check presence of Negative and SystemControl probes columns</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>negPrb_col <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">&quot;^Neg&quot;</span>, <span class="fu">colnames</span>(exprMat), <span class="at">value =</span> <span class="cn">TRUE</span>) <span class="co"># Neg columns</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>sysCon_col <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">&quot;^SystemControl&quot;</span>, <span class="fu">colnames</span>(exprMat), <span class="at">value =</span> <span class="cn">TRUE</span>) <span class="co"># SystemControl columns</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="co"># Show head expression matrix</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="fu">as.matrix</span>(exprMat[<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span></code></pre></div>
<pre><code>        Chrna4 Slc6a1 Cd109 Ldha Aldoc Drd1 Tank Rit2 Prkag2 Lpar3
c_1_1_0      9     58    11   30   938    6   28   20     15     2
c_1_1_1      0      0     0    0     0    0    0    0      0     0
c_1_1_2      0      0     0    0     0    0    0    0      1     0
c_1_1_3      5      1     0    1     2    0    0    0      0     0
c_1_1_4      1      1     0    1     0    0    0    2      1     0
c_1_1_5      0      0     0    2     2    0    0    0      0     0</code></pre>
<ul>
<li><strong>Dimensions:</strong> 48686 cells and 960 features.<br />
</li>
<li><strong>Cells of unassigned transcripts</strong>: 130 –&gt; c_1_1_0,
c_1_2_0, c_1_3_0, c_1_4_0, c_1_5_0, c_1_6_0, c_1_7_0, c_1_8_0, c_1_9_0,
c_1_10_0, c_1_11_0, c_1_12_0, c_1_13_0, c_1_14_0, c_1_15_0, c_1_16_0,
c_1_17_0, c_1_18_0, c_1_19_0, c_1_20_0, c_1_21_0, c_1_22_0, c_1_23_0,
c_1_24_0, c_1_25_0, c_1_26_0, c_1_27_0, c_1_28_0, c_1_29_0, c_1_30_0,
c_1_31_0, c_1_32_0, c_1_33_0, c_1_34_0, c_1_35_0, c_1_36_0, c_1_37_0,
c_1_38_0, c_1_39_0, c_1_40_0, c_1_41_0, c_1_42_0, c_1_43_0, c_1_44_0,
c_1_45_0, c_1_46_0, c_1_47_0, c_1_48_0, c_1_49_0, c_1_50_0, c_1_51_0,
c_1_52_0, c_1_53_0, c_1_54_0, c_1_55_0, c_1_56_0, c_1_57_0, c_1_58_0,
c_1_59_0, c_1_60_0, c_1_61_0, c_1_62_0, c_1_63_0, c_1_64_0, c_1_65_0,
c_1_66_0, c_1_67_0, c_1_68_0, c_1_69_0, c_1_70_0, c_1_71_0, c_1_72_0,
c_1_73_0, c_1_74_0, c_1_75_0, c_1_76_0, c_1_77_0, c_1_78_0, c_1_79_0,
c_1_80_0, c_1_81_0, c_1_82_0, c_1_83_0, c_1_84_0, c_1_85_0, c_1_86_0,
c_1_87_0, c_1_88_0, c_1_89_0, c_1_90_0, c_1_91_0, c_1_92_0, c_1_93_0,
c_1_94_0, c_1_95_0, c_1_96_0, c_1_97_0, c_1_98_0, c_1_99_0, c_1_100_0,
c_1_101_0, c_1_102_0, c_1_103_0, c_1_104_0, c_1_105_0, c_1_106_0,
c_1_107_0, c_1_108_0, c_1_109_0, c_1_110_0, c_1_111_0, c_1_112_0,
c_1_113_0, c_1_114_0, c_1_115_0, c_1_116_0, c_1_117_0, c_1_118_0,
c_1_119_0, c_1_120_0, c_1_121_0, c_1_122_0, c_1_123_0, c_1_124_0,
c_1_125_0, c_1_126_0, c_1_127_0, c_1_128_0, c_1_129_0, c_1_130_0.<br />
</li>
<li><strong>Negative probes:</strong> 10 –&gt; NegPrb1, NegPrb2,
NegPrb3, NegPrb4, NegPrb5, NegPrb6, NegPrb7, NegPrb8, NegPrb9,
NegPrb10.<br />
</li>
<li><strong>System Control probes:</strong> 0.</li>
</ul>
<!-- Note that, during the expression matrix loading with function "loadExprMat()", columns "fov" and "cell_ID", originally present in the raw file, have been removed and, instead, rows have been named with the combination: Slide_FOV_Cell number.   -->
<p>Although the number of rows/cells in the metadata and the expression
matrix should coincide, the expression matrix presents 130 extra rows.
This difference is due to the presence of cells of id “0”, which, in
CosMx SMI data, are used to store the transcripts counts that can not be
assigned to any cell (see an example <a
href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/assets/Pancreas-CosMx-ReadMe.html()">ReadMe
SMI Data File)</a>. Therefore, those extra rows represent the sum of
unassigned transcripts from the correspondent FOV and can be used to
evaluate segmentation.</p>
<p>Apart from that, some manipulation of the data is suspected, since
the negative probes, which are named “Negative” on the raw AtoMx output,
are found as “NegPrb”, and the System Control codes columns are not
present, suggesting its previous removal.</p>
</div>
</div>
<div id="format-the-data" class="section level1">
<h1>Format the data</h1>
<p>Before creating the Seurat object, some further formatting of the
data might be helpful:</p>
<div id="expression-matrix-formatting" class="section level2">
<h2>Expression Matrix formatting</h2>
<p>For further analysis, the expression matrix must only have the
interest features in columns and real cells in rows, with no additional
information.</p>
<p>As previously stated, rows with a “cell_ID” of 0 do not represent
actual cells, and false positive controls (Negative probes and System
Control codes) do not represent features of interest, therefore, they
must be removed from the expression matrix. However, they posses
interesting information that should be kept in the metadata for future
exploration.</p>
<!-- When exporting a Seurat object directly from AtoMx, the main expression matrix is kept as the "RNA" assay, while Negative probes and System Control codes represent two alternative assays, called "negprobes" and "falsecode", respectively. This pipeline will not create those assays, but will calculate their respective nCount and save them in the metadata. -->
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># Calculate number of transcripts not assigned to any cell (by FOV)</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>cells_0 <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">&quot;_0$&quot;</span>, <span class="fu">rownames</span>(exprMat)) <span class="co"># Rows with cell_id 0</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>negPrb_col <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">&quot;^NegPrb&quot;</span>, <span class="fu">colnames</span>(exprMat)) <span class="co"># NegPrb columns</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>unassigned <span class="ot">&lt;-</span> <span class="fu">data.table</span>(<span class="at">cell_id =</span> <span class="fu">character</span>(<span class="dv">0</span>), <span class="co"># Data.table to save the info</span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>                         <span class="at">fov =</span> <span class="fu">numeric</span>(<span class="dv">0</span>),</span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>                         <span class="at">unassigned_tx =</span> <span class="fu">numeric</span>(<span class="dv">0</span>))</span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a></span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> cells_0) {</span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a>  cell_id <span class="ot">&lt;-</span> <span class="fu">rownames</span>(exprMat)[i]</span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a>  fov <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">strsplit</span>(cell_id, <span class="st">&quot;_&quot;</span>)[[<span class="dv">1</span>]][<span class="dv">3</span>])</span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a>  unassigned_tx <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">rowSums</span>(exprMat[i, <span class="sc">-</span>negPrb_col, <span class="at">drop =</span> <span class="cn">FALSE</span>])) <span class="co"># Only real Tx</span></span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a>  </span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a>  new_row <span class="ot">&lt;-</span> <span class="fu">list</span>(cell_id, fov, unassigned_tx)</span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a>  unassigned <span class="ot">&lt;-</span> <span class="fu">rbindlist</span>(<span class="fu">list</span>(unassigned, new_row))</span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a>}</span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a></span>
<span id="cb10-18"><a href="#cb10-18" tabindex="-1"></a><span class="co"># Remove rows of unassigned transcripts from exprMat</span></span>
<span id="cb10-19"><a href="#cb10-19" tabindex="-1"></a>exprMat <span class="ot">&lt;-</span> exprMat[<span class="sc">-</span>cells_0,]</span>
<span id="cb10-20"><a href="#cb10-20" tabindex="-1"></a></span>
<span id="cb10-21"><a href="#cb10-21" tabindex="-1"></a><span class="co"># Calculate nCount of Negative and System Control probes</span></span>
<span id="cb10-22"><a href="#cb10-22" tabindex="-1"></a>sysCon_col <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">&quot;^SystemControl&quot;</span>, <span class="fu">colnames</span>(exprMat)) <span class="co"># SystemControl columns</span></span>
<span id="cb10-23"><a href="#cb10-23" tabindex="-1"></a></span>
<span id="cb10-24"><a href="#cb10-24" tabindex="-1"></a>falsePrb <span class="ot">&lt;-</span> <span class="fu">data.table</span>(<span class="at">cell_id =</span> <span class="fu">character</span>(<span class="dv">0</span>), <span class="co"># Data.table to save the info</span></span>
<span id="cb10-25"><a href="#cb10-25" tabindex="-1"></a>                       <span class="at">nCount_negprobes =</span> <span class="fu">numeric</span>(<span class="dv">0</span>),</span>
<span id="cb10-26"><a href="#cb10-26" tabindex="-1"></a>                       <span class="at">nCount_falsecode =</span> <span class="fu">numeric</span>(<span class="dv">0</span>))</span>
<span id="cb10-27"><a href="#cb10-27" tabindex="-1"></a></span>
<span id="cb10-28"><a href="#cb10-28" tabindex="-1"></a>nCount_negprobes <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(exprMat[, negPrb_col])</span>
<span id="cb10-29"><a href="#cb10-29" tabindex="-1"></a>nCount_falsecode <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(exprMat[, sysCon_col])</span>
<span id="cb10-30"><a href="#cb10-30" tabindex="-1"></a></span>
<span id="cb10-31"><a href="#cb10-31" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">identical</span>(<span class="fu">names</span>(nCount_falsecode), <span class="fu">names</span>(nCount_negprobes))){</span>
<span id="cb10-32"><a href="#cb10-32" tabindex="-1"></a>  cell_id <span class="ot">&lt;-</span> <span class="fu">names</span>(nCount_negprobes)</span>
<span id="cb10-33"><a href="#cb10-33" tabindex="-1"></a>  new_row <span class="ot">&lt;-</span> <span class="fu">list</span>(cell_id, nCount_negprobes, nCount_falsecode)</span>
<span id="cb10-34"><a href="#cb10-34" tabindex="-1"></a>  falsePrb <span class="ot">&lt;-</span> <span class="fu">rbindlist</span>(<span class="fu">list</span>(falsePrb, new_row))</span>
<span id="cb10-35"><a href="#cb10-35" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb10-36"><a href="#cb10-36" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">&quot;nCount_negprobes and nCount_falsecode don&#39;t share cell ids!&quot;</span>)</span>
<span id="cb10-37"><a href="#cb10-37" tabindex="-1"></a>}</span>
<span id="cb10-38"><a href="#cb10-38" tabindex="-1"></a></span>
<span id="cb10-39"><a href="#cb10-39" tabindex="-1"></a><span class="co"># Save Negative probes as a separate matrix (for future Insitutype)</span></span>
<span id="cb10-40"><a href="#cb10-40" tabindex="-1"></a>negMat <span class="ot">&lt;-</span> exprMat[, negPrb_col]</span>
<span id="cb10-41"><a href="#cb10-41" tabindex="-1"></a></span>
<span id="cb10-42"><a href="#cb10-42" tabindex="-1"></a><span class="co"># Remove Negative and SystemControl probes from exprMat</span></span>
<span id="cb10-43"><a href="#cb10-43" tabindex="-1"></a>exprMat <span class="ot">&lt;-</span> exprMat[, <span class="sc">-</span><span class="fu">c</span>(negPrb_col, sysCon_col)]</span></code></pre></div>
<ul>
<li><strong>Expression Matrix dimensions after formatting:</strong>
48556 cells and 950 features.</li>
</ul>
<p>After removing the cells with id “0”, the number of rows/cells match
that of the metadata. Also, the number of columns/features has been
diminish after removing the negative probes.</p>
</div>
<div id="fov-positions-formating" class="section level2">
<h2>FOV positions formating</h2>
<p>Depending on the AtoMx version from which the data was exported, this
file might contain the FOV coordinates in px (older versions) or in mm
(newer versions) (see <a
href="https://github.com/cancerbioinformatics/CosMx_Lite?tab=readme-ov-file#fov_position_filecsv">CosMxLite</a>).</p>
<p>The following code will check if the px coordinates are present or
not, and calculate them if needed.</p>
<!-- Conversion from mm to px has been obtained from CosMxLite vignette. However, this conversion might be different depending on resolution. -->
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="do">## Code adapted from CosMxLite vignette</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="co"># Check and/or calculate px coordinates</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">all</span>(<span class="fu">c</span>(<span class="st">&quot;x_global_px&quot;</span>, <span class="st">&quot;y_global_px&quot;</span>) <span class="sc">%in%</span> <span class="fu">colnames</span>(fov_positions))) {</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>  result <span class="ot">&lt;-</span> <span class="st">&quot;FOV coordinates in pixels are already present.&quot;</span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>  fov_positions<span class="sc">$</span>x_global_px <span class="ot">&lt;-</span> fov_positions<span class="sc">$</span>x_global_mm<span class="sc">*</span><span class="dv">1000</span><span class="sc">/</span><span class="fl">0.1202809</span> <span class="co"># converting mm to px </span></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>  fov_positions<span class="sc">$</span>y_global_px <span class="ot">&lt;-</span> fov_positions<span class="sc">$</span>y_global_mm<span class="sc">*</span><span class="dv">1000</span><span class="sc">/</span><span class="fl">0.1202809</span> <span class="co"># converting mm to px </span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a>  result <span class="ot">&lt;-</span> <span class="st">&quot;FOV coordinates in pixels have been calculated.&quot;</span></span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a>}</span></code></pre></div>
<p>FOV coordinates in pixels are already present.</p>
</div>
<div id="metadata-formating" class="section level2">
<h2>Metadata formating</h2>
<p>In some cases, the exported metadata file from AtoMx might contain
some extra information on analysis performed in said software. This
information increases the file size and might not be used when
performing the analysis from the beginning. Therefore, it can me removed
by selecting only the basic columns, or those of interest for your
analysis.</p>
<p>The following code adapted from <a
href="https://github.com/cancerbioinformatics/CosMx_Lite/tree/main/scripts">CosMxLite</a>
will subset the metadata by selecting the basic columns only.
Additionally, several new columns will be created to facilitate further
subsetting and visualizations:</p>
<!-- The following columns have been selected from experience, observing other datasets, as wells as the information observed when exporting data directly from AtoMx, which are not present in this dataset, but could be present in data from newer AtoMx versions. I have chosen the same names I have seen in said examples for clarity. -->
<ul>
<li><strong>x_FOV_px:</strong> including the X coordinates in pixels of
the FOV the cell resides in.<br />
</li>
<li><strong>y_FOV_px:</strong> including the Y coordinates in pixels of
the FOV the cell resides in.<br />
</li>
<li><strong>cell_id:</strong> including the slide_fov_cell number as a
cell unique identifier. This column will also be use to name the rows,
as previously done in the expression matrix.<br />
</li>
<li><strong>slide_ID_numeric:</strong> including the slide number.<br />
</li>
<li><strong>unassignedTranscripts:</strong> including the number of
unassigned transcripts of the FOV the cell resides in. This will be
re-calculated to fraction after creating the Seurat object.</li>
<li><strong>nCount_negprobes:</strong> including total Negative probes
counts per cell.<br />
</li>
<li><strong>nCount_falsecodes:</strong> including total System Control
probes counts per cell.</li>
</ul>
<p>Note, the “cell_ID” original column won’t be kept as it only provides
FOV-specific identification of the cells and might be misleading.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="do">## Code inspired by CosMxLite vignette</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="co"># Add new columns: slide_ID and cell_id</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>metadata[, slide_ID_numeric <span class="sc">:=</span> <span class="dv">1</span>]</span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a>slide_fov_cell <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;c_&quot;</span>, metadata<span class="sc">$</span>slide_ID_numeric, <span class="st">&quot;_&quot;</span>, metadata<span class="sc">$</span>fov, <span class="st">&quot;_&quot;</span>, metadata<span class="sc">$</span>cell_ID)</span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a>metadata[, cell_id <span class="sc">:=</span> slide_fov_cell]</span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a></span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a><span class="co"># Merge X and Y FOV coordinates by &quot;fov&quot;</span></span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a>metadata[fov_positions, on <span class="ot">=</span> <span class="st">&quot;fov&quot;</span>, <span class="st">`</span><span class="at">:=</span><span class="st">`</span> (<span class="at">x_FOV_px =</span> i.x_global_px,</span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a>                                          <span class="at">y_FOV_px =</span> i.y_global_px)]</span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a></span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a><span class="co"># Merge unassigned_tx by &quot;fov&quot;</span></span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a>metadata[unassigned, on <span class="ot">=</span> <span class="st">&quot;fov&quot;</span>, <span class="st">`</span><span class="at">:=</span><span class="st">`</span> (<span class="at">unassignedTranscripts =</span> i.unassigned_tx)]</span>
<span id="cb12-15"><a href="#cb12-15" tabindex="-1"></a></span>
<span id="cb12-16"><a href="#cb12-16" tabindex="-1"></a><span class="co"># Merge nCount_negprobes and nCount_falsecodes by &quot;cell_id&quot;</span></span>
<span id="cb12-17"><a href="#cb12-17" tabindex="-1"></a>metadata[falsePrb, on <span class="ot">=</span> <span class="st">&quot;cell_id&quot;</span>, <span class="st">`</span><span class="at">:=</span><span class="st">`</span> (<span class="at">nCount_negprobes =</span> i.nCount_negprobes,</span>
<span id="cb12-18"><a href="#cb12-18" tabindex="-1"></a>                                         <span class="at">nCount_falsecode =</span> i.nCount_falsecode)]</span>
<span id="cb12-19"><a href="#cb12-19" tabindex="-1"></a></span>
<span id="cb12-20"><a href="#cb12-20" tabindex="-1"></a><span class="co"># Aisle basic columns (from CosMxLite) + recently added columns </span></span>
<span id="cb12-21"><a href="#cb12-21" tabindex="-1"></a>columns_to_keep <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;slide_ID_numeric&quot;</span>, <span class="st">&quot;fov&quot;</span>, <span class="st">&quot;cell_id&quot;</span>, <span class="st">&quot;Area&quot;</span>, <span class="st">&quot;AspectRatio&quot;</span>,</span>
<span id="cb12-22"><a href="#cb12-22" tabindex="-1"></a>                     <span class="st">&quot;x_FOV_px&quot;</span>, <span class="st">&quot;y_FOV_px&quot;</span>, <span class="st">&quot;Width&quot;</span>, <span class="st">&quot;Height&quot;</span>, <span class="st">&quot;CenterX_local_px&quot;</span>,</span>
<span id="cb12-23"><a href="#cb12-23" tabindex="-1"></a>                     <span class="st">&quot;CenterY_local_px&quot;</span>, <span class="st">&quot;CenterX_global_px&quot;</span>, <span class="st">&quot;CenterY_global_px&quot;</span>,</span>
<span id="cb12-24"><a href="#cb12-24" tabindex="-1"></a>                     <span class="fu">colnames</span>(metadata)[<span class="fu">grep</span>(<span class="st">&quot;Mean|Max&quot;</span>, <span class="fu">colnames</span>(metadata))], </span>
<span id="cb12-25"><a href="#cb12-25" tabindex="-1"></a>                     <span class="st">&quot;unassignedTranscripts&quot;</span>, <span class="st">&quot;nCount_negprobes&quot;</span>, <span class="st">&quot;nCount_falsecode&quot;</span>)</span>
<span id="cb12-26"><a href="#cb12-26" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> metadata[, ..columns_to_keep]</span>
<span id="cb12-27"><a href="#cb12-27" tabindex="-1"></a></span>
<span id="cb12-28"><a href="#cb12-28" tabindex="-1"></a><span class="co"># Re-name rows as the slide_fov_cell</span></span>
<span id="cb12-29"><a href="#cb12-29" tabindex="-1"></a><span class="fu">rownames</span>(metadata) <span class="ot">&lt;-</span> metadata<span class="sc">$</span>cell_id</span>
<span id="cb12-30"><a href="#cb12-30" tabindex="-1"></a></span>
<span id="cb12-31"><a href="#cb12-31" tabindex="-1"></a><span class="co"># Ensure that cell content and order in the ExprMat and metadata matches</span></span>
<span id="cb12-32"><a href="#cb12-32" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">vsetequal</span>(<span class="fu">rownames</span>(metadata), <span class="fu">rownames</span>(exprMat))) {</span>
<span id="cb12-33"><a href="#cb12-33" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">&quot;Rows do not match between metadata and exprMat.&quot;</span>)</span>
<span id="cb12-34"><a href="#cb12-34" tabindex="-1"></a>} <span class="cf">else</span> <span class="cf">if</span> (<span class="sc">!</span><span class="fu">identical</span>(<span class="fu">rownames</span>(metadata), <span class="fu">rownames</span>(exprMat))) {</span>
<span id="cb12-35"><a href="#cb12-35" tabindex="-1"></a>  exprMat <span class="ot">&lt;-</span> exprMat[<span class="fu">match</span>(<span class="fu">rownames</span>(metadata), <span class="fu">rownames</span>(exprMat)), ]</span>
<span id="cb12-36"><a href="#cb12-36" tabindex="-1"></a>}</span></code></pre></div>
<ul>
<li><strong>Metadata columns after formatting:</strong>
slide_ID_numeric, fov, cell_id, Area, AspectRatio, x_FOV_px, y_FOV_px,
Width, Height, CenterX_local_px, CenterY_local_px, CenterX_global_px,
CenterY_global_px, Mean.Histone, Max.Histone, Mean.G, Max.G, Mean.rRNA,
Max.rRNA, Mean.GFAP, Max.GFAP, Mean.DAPI, Max.DAPI,
unassignedTranscripts, nCount_negprobes, nCount_falsecode.</li>
</ul>
</div>
<div id="save-intermediate-data" class="section level2">
<h2>Save intermediate data</h2>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="do">## Code adapted from Scratch Space vignette</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;processed_data&quot;</span>))){</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;processed_data&quot;</span>))</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>}</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a></span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a><span class="fu">saveRDS</span>(exprMat, <span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;processed_data&quot;</span>,<span class="st">&quot;exprMat_unfiltered.RDS&quot;</span>))</span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a><span class="fu">saveRDS</span>(metadata, <span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;processed_data&quot;</span>,<span class="st">&quot;metadata_unfiltered.RDS&quot;</span>))</span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a><span class="fu">saveRDS</span>(fov_positions, <span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;processed_data&quot;</span>,<span class="st">&quot;fov_positions_unfiltered.RDS&quot;</span>))</span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a><span class="fu">saveRDS</span>(negMat, <span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;processed_data&quot;</span>,<span class="st">&quot;negMat_unfiltered.RDS&quot;</span>))</span></code></pre></div>
</div>
</div>
<div id="create-a-seurat-object" class="section level1">
<h1>Create a Seurat object</h1>
<p>The Seurat package has two built-in functions to read the flat files
from CosMx:</p>
<ul>
<li><strong>ReadNanostring:</strong> this function reads the available
flat files and produces a list with the expression matrix, the metadata,
the centroids, etc.<br />
</li>
<li><strong>LoadNanostring:</strong> this function reads the available
flat files and produces a Seurat object. However, for this function to
work, all the files (including tx_file and polygons) must be
available.</li>
</ul>
<p>In this pipeline, the Seurat object will be created with the
“CreateSeuratObject” function, as seen in the <a
href="https://github.com/cancerbioinformatics/CosMx_Lite/tree/main/scripts">CosMxLite
vignette</a>, using the data previously loaded and processed.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="do">## Code adapted from CosMxLite vignette</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="co"># Create a basic Seurat object from the processed exprMat and metadata</span></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a>seu <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> <span class="fu">t</span>(exprMat), <span class="at">project =</span> <span class="st">&quot;CosMx&quot;</span>,</span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a>                          <span class="at">meta.data =</span> metadata, <span class="at">min.cells =</span> <span class="dv">3</span>, <span class="at">min.features =</span> <span class="dv">1</span>)</span></code></pre></div>
<!-- Note I have filtered by min.cells and min.features as seen in CosMxLite -->
<div id="calculate-fraction-of-unassigned-transcripts-per-fov"
class="section level2">
<h2>Calculate fraction of unassigned transcripts per FOV</h2>
<p>When creating a Seurat object, “nCount” and “nFeature” columns are
automatically calculated and added to the meta.data slot. Now, the
“unassignedTranscripts” column can be recalculated to represent the
fraction of total transcripts.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="co"># Calculate total transcripts per FOV and add to meta.data</span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>nCount_FOV <span class="ot">&lt;-</span> seu<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span></span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>  <span class="fu">group_by</span>(fov) <span class="sc">%&gt;%</span></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">nCount_FOV =</span> <span class="fu">sum</span>(nCount_RNA))</span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a>seu<span class="sc">@</span>meta.data <span class="ot">&lt;-</span> seu<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span></span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a>  <span class="fu">left_join</span>(nCount_FOV, <span class="at">by =</span> <span class="st">&quot;fov&quot;</span>)</span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a></span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a><span class="co"># Re-assign rownames to metadata, lost by dplyr functions</span></span>
<span id="cb15-10"><a href="#cb15-10" tabindex="-1"></a><span class="fu">rownames</span>(seu<span class="sc">@</span>meta.data) <span class="ot">&lt;-</span> seu<span class="sc">$</span>cell_id</span>
<span id="cb15-11"><a href="#cb15-11" tabindex="-1"></a></span>
<span id="cb15-12"><a href="#cb15-12" tabindex="-1"></a><span class="co"># Re-calculate unassignedTranscripts to fraction through dividing by</span></span>
<span id="cb15-13"><a href="#cb15-13" tabindex="-1"></a><span class="co"># total detected transcripts: nCount_FOV + unassignedTranscripts</span></span>
<span id="cb15-14"><a href="#cb15-14" tabindex="-1"></a>seu<span class="sc">$</span>unassignedTranscripts <span class="ot">&lt;-</span> seu<span class="sc">$</span>unassignedTranscripts <span class="sc">/</span> (seu<span class="sc">$</span>nCount_FOV <span class="sc">+</span> seu<span class="sc">$</span>unassignedTranscripts)</span></code></pre></div>
</div>
<div id="add-spatial-coordinates" class="section level2">
<h2>Add Spatial coordinates</h2>
<p>Finally, it is important to include cell coordinates for
visualizations. In the <a
href="https://github.com/cancerbioinformatics/CosMx_Lite/tree/main/scripts">CosMxLite
vignette</a> two alternatives on adding the spatial coordinates are
shown:</p>
<ul>
<li><strong>Add as “global FOV”:</strong> this method utilizes the
“images” slot of the Seurat object and the “CreateCentroids” and
“CreateFOV” functions, just as it is done in the Seurat function
“LoadNanostring”. By using this method, the resulting structure is the
same as in the AtoMx exported Seurat object.<br />
</li>
<li><strong>Add as a “dimensional reduction”:</strong> this method
utilizes the “reductions” slot of the Seurat object and the
“CreateDimReducObject” function. By using this method, more
visualization options are available.</li>
</ul>
<p>In this pipeline, the “global FOV” methods will be used to maintain
interoperability with the AtoMx Seurat output.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="do">## Code adapted from CosMxLite vignette: in this vignette two alternatives</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="do">## on adding the spatial coordinates are shown.</span></span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a><span class="co"># Add cell coordinates as a &quot;global FOV&quot;</span></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a></span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a><span class="co"># Note: x and y coordinates need to be swapped to orient the tissue correctly</span></span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a>cell_centroids <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">y =</span> seu<span class="sc">$</span>CenterX_global_px, </span>
<span id="cb16-8"><a href="#cb16-8" tabindex="-1"></a>                             <span class="at">x =</span> seu<span class="sc">$</span>CenterY_global_px, </span>
<span id="cb16-9"><a href="#cb16-9" tabindex="-1"></a>                             <span class="at">cell =</span> <span class="fu">colnames</span>(seu))</span>
<span id="cb16-10"><a href="#cb16-10" tabindex="-1"></a>cents <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="st">&quot;centroids&quot;</span> <span class="ot">=</span> <span class="fu">CreateCentroids</span>(cell_centroids))</span>
<span id="cb16-11"><a href="#cb16-11" tabindex="-1"></a>coords <span class="ot">&lt;-</span> <span class="fu">CreateFOV</span>(<span class="at">coords =</span> cents, <span class="at">type =</span> <span class="fu">c</span>(<span class="st">&quot;centroids&quot;</span>), <span class="at">assay =</span> <span class="st">&quot;RNA&quot;</span>)</span>
<span id="cb16-12"><a href="#cb16-12" tabindex="-1"></a>seu[[<span class="st">&quot;globalFOV&quot;</span>]] <span class="ot">&lt;-</span> coords</span>
<span id="cb16-13"><a href="#cb16-13" tabindex="-1"></a></span>
<span id="cb16-14"><a href="#cb16-14" tabindex="-1"></a><span class="co"># Add cell coordinates as a &quot;reduction&quot; (left as an example)</span></span>
<span id="cb16-15"><a href="#cb16-15" tabindex="-1"></a><span class="co"># coordinates &lt;- as.matrix(data.frame(CosMx_1 = seu$CenterX_global_px, </span></span>
<span id="cb16-16"><a href="#cb16-16" tabindex="-1"></a><span class="co">#                                     CosMx_2 = seu$CenterY_global_px))</span></span>
<span id="cb16-17"><a href="#cb16-17" tabindex="-1"></a><span class="co"># seu[[&#39;Centroids&#39;]] &lt;- CreateDimReducObject(embeddings = coordinates, key = &quot;CosMx_&quot;, </span></span>
<span id="cb16-18"><a href="#cb16-18" tabindex="-1"></a><span class="co">#                                            global = TRUE, assay = &quot;RNA&quot;)</span></span></code></pre></div>
<!-- If transcripts coordinates from tx_file want to be included for future analysis, the appropriate slot to include them in would be: seu@images[["globalFOV"]]@molecules -->
</div>
<div id="save-the-seurat-object" class="section level2">
<h2>Save the Seurat object</h2>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="co"># Save the Seurat object</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="fu">saveRDS</span>(seu, <span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;processed_data&quot;</span>,<span class="st">&quot;seu_semifiltered.RDS&quot;</span>))</span></code></pre></div>
</div>
</div>
<div id="performance-and-session-info" class="section level1">
<h1>Performance and Session Info</h1>
<details>
<summary>
<strong>Performance Report</strong>
</summary>
<table class="table table-striped table-hover table-condensed" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
Chunk
</th>
<th style="text-align:right;">
Time_sec
</th>
<th style="text-align:right;">
Memory_Mb
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Libraries
</td>
<td style="text-align:right;">
1.47
</td>
<td style="text-align:right;">
148.8
</td>
</tr>
<tr>
<td style="text-align:left;">
LocateFiles
</td>
<td style="text-align:right;">
0.13
</td>
<td style="text-align:right;">
0.0
</td>
</tr>
<tr>
<td style="text-align:left;">
LoadSlide_1
</td>
<td style="text-align:right;">
3.30
</td>
<td style="text-align:right;">
219.4
</td>
</tr>
<tr>
<td style="text-align:left;">
ExploringFOVpositions
</td>
<td style="text-align:right;">
0.13
</td>
<td style="text-align:right;">
0.8
</td>
</tr>
<tr>
<td style="text-align:left;">
ExploringMetadata
</td>
<td style="text-align:right;">
0.14
</td>
<td style="text-align:right;">
0.2
</td>
</tr>
<tr>
<td style="text-align:left;">
ExploringExprMat
</td>
<td style="text-align:right;">
0.15
</td>
<td style="text-align:right;">
0.0
</td>
</tr>
<tr>
<td style="text-align:left;">
PreparingExprMat
</td>
<td style="text-align:right;">
3.49
</td>
<td style="text-align:right;">
2.8
</td>
</tr>
<tr>
<td style="text-align:left;">
PreparingFOVpositions
</td>
<td style="text-align:right;">
0.13
</td>
<td style="text-align:right;">
0.0
</td>
</tr>
<tr>
<td style="text-align:left;">
PreparingMetadata
</td>
<td style="text-align:right;">
0.72
</td>
<td style="text-align:right;">
2.8
</td>
</tr>
<tr>
<td style="text-align:left;">
SavingData
</td>
<td style="text-align:right;">
12.99
</td>
<td style="text-align:right;">
0.1
</td>
</tr>
<tr>
<td style="text-align:left;">
CreateSeuObj
</td>
<td style="text-align:right;">
2.46
</td>
<td style="text-align:right;">
215.5
</td>
</tr>
<tr>
<td style="text-align:left;">
UnassignedTxFrac
</td>
<td style="text-align:right;">
0.19
</td>
<td style="text-align:right;">
3.4
</td>
</tr>
<tr>
<td style="text-align:left;">
Coords
</td>
<td style="text-align:right;">
0.40
</td>
<td style="text-align:right;">
4.7
</td>
</tr>
<tr>
<td style="text-align:left;">
SavingSeuObj
</td>
<td style="text-align:right;">
18.37
</td>
<td style="text-align:right;">
0.0
</td>
</tr>
<tr>
<td style="text-align:left;">
Total
</td>
<td style="text-align:right;">
44.07
</td>
<td style="text-align:right;">
598.5
</td>
</tr>
</tbody>
</table>
</details>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>R version 4.4.3 (2025-02-28 ucrt)
Platform: x86_64-w64-mingw32/x64
Running under: Windows 10 x64 (build 19045)

Matrix products: default


locale:
[1] LC_COLLATE=Spanish_Spain.utf8  LC_CTYPE=Spanish_Spain.utf8   
[3] LC_MONETARY=Spanish_Spain.utf8 LC_NUMERIC=C                  
[5] LC_TIME=Spanish_Spain.utf8    

time zone: Europe/Madrid
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] SeuratObject_4.1.4 Seurat_4.4.0       kableExtra_1.4.0   dplyr_1.1.4       
[5] vecsets_1.4        here_1.0.1         Matrix_1.7-2       data.table_1.17.0 
[9] workflowr_1.7.1   

loaded via a namespace (and not attached):
  [1] deldir_2.0-4           pbapply_1.7-4          gridExtra_2.3         
  [4] rlang_1.1.4            magrittr_2.0.3         git2r_0.36.2          
  [7] RcppAnnoy_0.0.22       spatstat.geom_3.5-0    matrixStats_1.4.1     
 [10] ggridges_0.5.6         compiler_4.4.3         getPass_0.2-4         
 [13] reshape2_1.4.4         png_0.1-8              systemfonts_1.1.0     
 [16] callr_3.7.6            vctrs_0.6.5            stringr_1.5.1         
 [19] pkgconfig_2.0.3        fastmap_1.2.0          promises_1.3.2        
 [22] rmarkdown_2.29         pracma_2.4.4           ps_1.8.1              
 [25] purrr_1.0.2            xfun_0.49              cachem_1.1.0          
 [28] jsonlite_1.8.9         goftest_1.2-3          later_1.4.1           
 [31] spatstat.utils_3.1-5   irlba_2.3.5.1          parallel_4.4.3        
 [34] cluster_2.1.8          R6_2.5.1               ica_1.0-3             
 [37] spatstat.data_3.1-6    bslib_0.8.0            stringi_1.8.4         
 [40] RColorBrewer_1.1-3     reticulate_1.40.0      spatstat.univar_3.1-4 
 [43] parallelly_1.41.0      scattermore_1.2        lmtest_0.9-40         
 [46] jquerylib_0.1.4        Rcpp_1.0.13-1          knitr_1.49            
 [49] tensor_1.5.1           future.apply_1.11.3    zoo_1.8-14            
 [52] sctransform_0.4.2      httpuv_1.6.15          splines_4.4.3         
 [55] igraph_2.1.2           tidyselect_1.2.1       abind_1.4-8           
 [58] rstudioapi_0.17.1      yaml_2.3.10            spatstat.random_3.4-1 
 [61] spatstat.explore_3.5-2 codetools_0.2-20       miniUI_0.1.1.1        
 [64] processx_3.8.4         listenv_0.9.1          plyr_1.8.9            
 [67] lattice_0.22-6         tibble_3.2.1           shiny_1.10.0          
 [70] ROCR_1.0-11            evaluate_1.0.1         Rtsne_0.17            
 [73] future_1.34.0          survival_3.8-3         polyclip_1.10-7       
 [76] xml2_1.3.6             fitdistrplus_1.2-4     pillar_1.10.0         
 [79] whisker_0.4.1          KernSmooth_2.23-26     plotly_4.11.0         
 [82] generics_0.1.3         rprojroot_2.0.4        sp_2.2-0              
 [85] ggplot2_3.5.1          munsell_0.5.1          scales_1.3.0          
 [88] globals_0.16.3         xtable_1.8-4           glue_1.8.0            
 [91] lazyeval_0.2.2         tools_4.4.3            RANN_2.6.2            
 [94] fs_1.6.5               leiden_0.4.3.1         cowplot_1.1.3         
 [97] grid_4.4.3             tidyr_1.3.1            colorspace_2.1-1      
[100] nlme_3.1-167           patchwork_1.3.0        cli_3.6.3             
[103] spatstat.sparse_3.1-0  viridisLite_0.4.2      svglite_2.1.3         
[106] uwot_0.2.3             gtable_0.3.6           sass_0.4.9            
[109] digest_0.6.37          progressr_0.15.1       ggrepel_0.9.6         
[112] htmlwidgets_1.6.4      farver_2.1.2           htmltools_0.5.8.1     
[115] lifecycle_1.0.4        httr_1.4.7             mime_0.12             
[118] MASS_7.3-64           </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
