<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>CosMx Data Insitutype Unsupervised Clustering and scType annotation</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #204a87; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #204a87; font-weight: bold; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #ce5c00; font-weight: bold; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>







<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">CosMx™ SMI Data Analysis Pipeline</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="dataset.html">Dataset</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pipeline
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="pipeline.html">Introduction</a>
    </li>
    <li>
      <a href="0.0_data_loading.html">Data Loading</a>
    </li>
    <li>
      <a href="1.0_qc_and_filtering.html">QC and filtering</a>
    </li>
    <li>
      <a href="2.0_normalization.html">Normalization</a>
    </li>
    <li>
      <a href="3.0_dimensional_reduction.html">Dimensional reduction</a>
    </li>
    <li>
      <a href="4.0_insitutype_cell_typing.html">Cell typing</a>
    </li>
  </ul>
</li>
<li>
  <a href="norm_examples.html">Normalization</a>
</li>
<li>
  <a href="clus_examples.html">Cell typing</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
<li>
  <a href="https://github.com/lga-uoc/CosMx_pipeline_LGA">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">CosMx Data Insitutype Unsupervised
Clustering and scType annotation</h1>
<h4 class="author">LGA</h4>
<h4 class="date">03 septiembre 2025</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-09-03
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>CosMx_pipeline_LGA/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20250517code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20250517)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20250517code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20250517)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomlgauocCosMxpipelineLGAtreecc8470c609588ae6794c96ed3c178f92b6f9c1fdtargetblankcc8470ca">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/lga-uoc/CosMx_pipeline_LGA/tree/cc8470c609588ae6794c96ed3c178f92b6f9c1fd" target="_blank">cc8470c</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomlgauocCosMxpipelineLGAtreecc8470c609588ae6794c96ed3c178f92b6f9c1fdtargetblankcc8470ca"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/lga-uoc/CosMx_pipeline_LGA/tree/cc8470c609588ae6794c96ed3c178f92b6f9c1fd" target="_blank">cc8470c</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    NBClust-Plots/
    Ignored:    analysis/.Rhistory
    Ignored:    data/flatFiles/CoronalHemisphere/Run1000_S1_Half_exprMat_file.csv
    Ignored:    data/flatFiles/CoronalHemisphere/Run1000_S1_Half_fov_positions_file.csv
    Ignored:    data/flatFiles/CoronalHemisphere/Run1000_S1_Half_metadata_file.csv
    Ignored:    output/processed_data/Log/
    Ignored:    output/processed_data/RC/
    Ignored:    output/processed_data/SCT/
    Ignored:    output/processed_data/exprMat_unfiltered.RDS
    Ignored:    output/processed_data/fov_positions_unfiltered.RDS
    Ignored:    output/processed_data/metadata_unfiltered.RDS
    Ignored:    output/processed_data/negMat_unfiltered.RDS
    Ignored:    output/processed_data/seu_filtered.RDS
    Ignored:    output/processed_data/seu_semifiltered.RDS

Unstaged changes:
    Modified:   CosMx_pipeline_LGA.Rproj
    Modified:   output/performance_reports/0.0_data_loading_PR.csv
    Modified:   output/performance_reports/1.0_qc_and_filtering_PR.csv
    Modified:   output/performance_reports/2.0_normalization_PR.csv
    Modified:   output/performance_reports/3.0_dimensional_reduction_PR.csv
    Modified:   output/performance_reports/4.0_insitutype_cell_typing_PR.csv
    Deleted:    output/performance_reports/4.1_insitutype_unsup_clustering_PR.csv
    Deleted:    output/performance_reports/4.2_seurat_unsup_clustering_PR.csv
    Deleted:    output/performance_reports/5.0_RC_normalization_PR.csv
    Deleted:    output/performance_reports/5.1_RC_dimensional_reduction_PR.csv
    Deleted:    output/performance_reports/6.0_Log_normalization_PR.csv
    Deleted:    output/performance_reports/6.1_Log_dimensional_reduction_PR.csv
    Deleted:    output/performance_reports/pipeline_PR.csv

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>analysis/4.1_insitutype_unsup_clustering.Rmd</code>) and HTML
(<code>docs/4.1_insitutype_unsup_clustering.html</code>) files. If
you’ve configured a remote Git repository (see
<code>?wflow_git_remote</code>), click on the hyperlinks in the table
below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/lga-uoc/CosMx_pipeline_LGA/43970ac7eece0aded282c805767d439617b6a2a0/docs/4.1_insitutype_unsup_clustering.html" target="_blank">43970ac</a>
</td>
<td>
lga-uoc
</td>
<td>
2025-09-02
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/lga-uoc/CosMx_pipeline_LGA/blob/78c8449fe685bde2d835ac647c9135d949e8d3e2/analysis/4.1_insitutype_unsup_clustering.Rmd" target="_blank">78c8449</a>
</td>
<td>
lga-uoc
</td>
<td>
2025-09-02
</td>
<td>
Add Rmds files in analysis
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<!-- This pipeline has been developed using Seurat v4.4, as Seurat v5 presents several changes to the object structure that affected the working methods. Aditionally, the AtoMx exported Seurat objects come in the Seurat v4 structure. Therefore, working in this version can be more compatible with exported data too. -->
<div id="dependencies" class="section level1">
<h1>Dependencies</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(data.table) <span class="co"># Efficient data management</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(Matrix) <span class="co"># Sparse matrices</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">library</span>(here) <span class="co"># Enhanced file referencing in project-oriented workflows</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">library</span>(dplyr) <span class="co"># For the use of pipes %&gt;%</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="fu">library</span>(kableExtra) <span class="co"># For table formatting</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="fu">library</span>(Seurat) <span class="co"># Seurat object</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="fu">library</span>(ggplot2) <span class="co"># Graphics</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="fu">library</span>(patchwork) <span class="co"># Layout graphics</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="fu">library</span>(pheatmap) <span class="co"># Heatmap plot</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="fu">library</span>(HGNChelper) <span class="co"># For scType annotation</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="co">#remotes::install_github(&quot;Nanostring-Biostats/InSituType&quot;)</span></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="fu">library</span>(InSituType)</span></code></pre></div>
</div>
<div id="load-the-data" class="section level1">
<h1>Load the data</h1>
<p>First of all, data needs to be loaded into the session. For this
script, two inputs will be needed:</p>
<ul>
<li>The Seurat object from the “Dimensional reduction” script.<br />
</li>
<li>The Negative probe matrix from the “Data loading” script.</li>
</ul>
<!-- Insitutype algorithm uses not normalized raw counts to cluster the cells. However, the UMAP reduction will be used for plot visualizations, and is therefore needed. For the main pipeline I will use the SCT Seurat object. -->
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Indicate the object folder</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>folder <span class="ot">&lt;-</span> <span class="st">&quot;SCT&quot;</span> <span class="co"># Choose between &quot;RC&quot;, &quot;Log&quot; or &quot;SCT&quot;</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span>folder <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;RC&quot;</span>, <span class="st">&quot;Log&quot;</span>, <span class="st">&quot;SCT&quot;</span>)) {</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">&quot;The selected folder is invalid, choose: &#39;RC&#39;, &#39;Log&#39; or &#39;SCT&#39;&quot;</span>)</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>}</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co"># Load Seurat object</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>name <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;seu_&quot;</span>, folder, <span class="st">&quot;_um.RDS&quot;</span>)</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>seu <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;processed_data&quot;</span>,folder,name))</span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="co"># Load negMat</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a>negcounts <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;processed_data&quot;</span>,<span class="st">&quot;negMat_unfiltered.RDS&quot;</span>))</span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="co"># Load the number of PCs used for dimensional reduction</span></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a>name <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;npcs_&quot;</span>, folder, <span class="st">&quot;.RDS&quot;</span>)</span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a>npcs <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;processed_data&quot;</span>,folder,name))</span></code></pre></div>
<p>In this script, clustering will be done using the unsupervised
clustering method by <em>InSituType</em>. This method was developed by
Nanostring, owner of the CosMx SMI technology, and is, therefore,
optimized to work with this type of data specifically.</p>
</div>
<div id="prepare-the-data" class="section level1">
<h1>Prepare the data</h1>
<p>In order to apply the InSituType algorithm, it is necessary to
provide the function with the following information:</p>
<ul>
<li>Raw expression counts matrix (not normalized), with cells in columns
and genes in rows.<br />
</li>
<li>A vector of the mean negative control value per cell.</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Extract raw counts from the Seurat object</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">t</span>(seu<span class="sc">@</span>assays[[<span class="st">&quot;RNA&quot;</span>]]<span class="sc">@</span>counts)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co"># Note the matrix is transposed to have cells in rows and genes in cols</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co"># Ensure negcounts has the same cells as counts</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">identical</span>(<span class="fu">rownames</span>(counts), <span class="fu">rownames</span>(negcounts))) {</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>    negcounts <span class="ot">&lt;-</span> negcounts[<span class="fu">rownames</span>(counts),] <span class="co"># Filter and/or re-order if needed</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>} </span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a><span class="co"># Calculate Negative control means per cell</span></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>negmean <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(negcounts)</span></code></pre></div>
<p>Additionally, the function can take the following optional inputs, if
available:</p>
<ul>
<li>A vector of background counts per gene in each cell –&gt;
<em>Insitutype can automatically estimate this with the negative control
counts</em>.<br />
</li>
<li>A reference profile matrix with genes in rows and cell types in
columns –&gt; <em>This is only necessary for supervised or
semi-supervised clustering</em>.<br />
</li>
<li>Additional information incorporated as “cohort”, such as
inmunofluorescence stains or precomputed spatial niches –&gt;
<em>Including this option is low-risk, as it will not be taken into
account if it is uninformative</em>.</li>
</ul>
<p>For this example, immunofluoresce data will be included as a
cohort.</p>
<!-- In case no cohort wants to be included, this chunk can be commented so that it is not executed -->
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Extract Immunoflurence data (seen in Alex Soupir video&#39;s)</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>imm_fluor <span class="ot">&lt;-</span> seu<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">contains</span>(<span class="fu">c</span>(<span class="st">&quot;Mean&quot;</span>,<span class="st">&quot;Max&quot;</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="fu">contains</span>(<span class="st">&quot;DAPI&quot;</span>))</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co"># Set as cohort with InSituType function &quot;fastCohorting&quot;</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>cohort <span class="ot">&lt;-</span> <span class="fu">fastCohorting</span>(imm_fluor, <span class="at">gaussian_transform =</span> <span class="cn">TRUE</span>)</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co"># This function creates &quot;groups&quot; or similar cells based on their immunofluoresce,</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="co"># as that is the input I have given it</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="fu">table</span>(cohort)</span></code></pre></div>
<pre><code>cohort
    1     2     3     4     5     6     7     8     9    10 
 5157 12430 10502  2766  4056  3063  2531  1598  3138  2677 </code></pre>
<p>As it can be observed, the “fastCohorting” function has created
“groups” of similar cells based on their immunofluoresce abundances.</p>
</div>
<div id="unsupervised-insitutype-clustering" class="section level1">
<h1>Unsupervised Insitutype clustering</h1>
<p>In this method, the user provides a number or a range of clusters to
evaluate. Then the function will cluster the cells on the selected
number of clusters.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="do">## Code adapted from the InSituType vignette</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co"># Unsupervised Insitutye clustering with immunofluoresce cohort</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>unsup <span class="ot">&lt;-</span> <span class="fu">insitutype</span>(</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>  <span class="at">x =</span> counts,</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>  <span class="at">neg =</span> negmean,</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>  <span class="at">assay_type =</span> <span class="st">&quot;RNA&quot;</span>,</span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>  <span class="at">cohort =</span> cohort,</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a>  <span class="at">bg =</span> <span class="cn">NULL</span>, <span class="co"># The function will estimate it automatically</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a>  <span class="at">reference_profiles =</span> <span class="cn">NULL</span>, <span class="co"># Not needed for unsupervised clustering</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a>  <span class="at">n_clusts =</span> <span class="dv">5</span><span class="sc">:</span><span class="dv">20</span>, <span class="co"># Will test all these clusters and select the better fit</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a>  <span class="at">n_phase1 =</span> <span class="dv">200</span>,</span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a>  <span class="at">n_phase2 =</span> <span class="dv">500</span>,</span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a>  <span class="at">n_phase3 =</span> <span class="dv">2000</span>,</span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a>  <span class="at">n_starts =</span> <span class="dv">1</span>,</span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a>  <span class="at">max_iters =</span> <span class="dv">5</span></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a>)</span></code></pre></div>
<p>As a result the function provides a list with:</p>
<ul>
<li>Vector of cluster assignments.<br />
</li>
<li>Vector of probabilities, or confidence, of each cell’s
cluster.<br />
</li>
<li>Matrix of log-likelihoods with cells on rows and clusters in
columns.<br />
</li>
<li>Matrix of cluster-specific expression profiles.</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="do">## Code adapted from the InSituType vignette</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="co"># Clusters</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>n_clus <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">unique</span>(unsup<span class="sc">$</span>clust))</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a><span class="co"># Example of probabilities</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">head</span>(unsup<span class="sc">$</span>prob), <span class="dv">2</span>)</span></code></pre></div>
<pre><code>c_1_1_1 c_1_1_2 c_1_1_3 c_1_1_4 c_1_1_5 c_1_1_6 
      1       1       1       1       1       1 </code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">summary</span>(unsup<span class="sc">$</span>prob)</span></code></pre></div>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 0.3271  1.0000  1.0000  0.9895  1.0000  1.0000 </code></pre>
<p>As it can be observed, the algorithm has selected 17 clusters as the
better fit, with the majority of cells presenting a high confidence of
belonging to the selected cluster.</p>
<p>Now, it might be interesting to explore the confidence ratios in
order to select an appropriate threshold for low probability
assignments, and/or see if the low probabilities are concentrated in any
particular cluster.</p>
<div id="exploration-of-cell-assignment-scores" class="section level2">
<h2>Exploration of cell assignment scores</h2>
<p>As per the <a
href="https://github.com/Nanostring-Biostats/InSituType/blob/main/vignettes/NSCLC-clustering-vignette.Rmd">InSituType
vignette</a>, clusters with an average probability of less than 75% are
unreliable. Therefore, for the exploration a threshold of 75% will be
selected.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="co"># Determine initial low confidence cells</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>thresh <span class="ot">&lt;-</span> <span class="fl">0.75</span> </span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>low_conf_cells <span class="ot">&lt;-</span> <span class="fu">names</span>(unsup<span class="sc">$</span>prob[unsup<span class="sc">$</span>prob <span class="sc">&lt;</span> thresh])</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>n_low <span class="ot">&lt;-</span> <span class="fu">length</span>(low_conf_cells)</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>perc_low <span class="ot">&lt;-</span> <span class="fu">round</span>(n_low <span class="sc">/</span> <span class="fu">length</span>(unsup<span class="sc">$</span>prob) <span class="sc">*</span> <span class="dv">100</span>, <span class="dv">1</span>)</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a><span class="co"># Create dataframe for plotting</span></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">clust =</span> unsup<span class="sc">$</span>clust,</span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a>                 <span class="at">prob =</span> unsup<span class="sc">$</span>prob)</span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="co"># Histogram of confidence</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> prob)) <span class="sc">+</span></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">50</span>, <span class="at">fill =</span> <span class="st">&quot;lightblue&quot;</span>, <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> thresh, <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>, <span class="at">lwd =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="fu">paste</span>(<span class="st">&quot;Nº cells with prob &lt;&quot;</span>, thresh, <span class="st">&quot;:&quot;</span>, n_low)) <span class="sc">+</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">paste</span>(<span class="st">&quot;Nº cells with prob &lt;&quot;</span>, thresh, <span class="st">&quot;:&quot;</span>, n_low),</span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">&quot;Assignment Confidence (%)&quot;</span>,</span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&quot;Nº Cells&quot;</span>)</span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a></span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a><span class="co"># Boxplots of confidence by cluster</span></span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> clust, <span class="at">y =</span> prob)) <span class="sc">+</span></span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> thresh, <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>, <span class="at">lwd =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Confidence per Cluster&quot;</span>,</span>
<span id="cb12-15"><a href="#cb12-15" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">&quot;Cluster&quot;</span>, </span>
<span id="cb12-16"><a href="#cb12-16" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&quot;Assignment Confidence (%)&quot;</span>)</span>
<span id="cb12-17"><a href="#cb12-17" tabindex="-1"></a></span>
<span id="cb12-18"><a href="#cb12-18" tabindex="-1"></a><span class="co"># Arrange plots</span></span>
<span id="cb12-19"><a href="#cb12-19" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code></pre></div>
<p><img src="figure/4.1_insitutype_unsup_clustering.Rmd/VizUnsupProb-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-VizUnsupProb-1">
Past versions of VizUnsupProb-1.png
</button>
</p>
<div id="fig-VizUnsupProb-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/lga-uoc/CosMx_pipeline_LGA/blob/43970ac7eece0aded282c805767d439617b6a2a0/docs/figure/4.1_insitutype_unsup_clustering.Rmd/VizUnsupProb-1.png" target="_blank">43970ac</a>
</td>
<td>
lga-uoc
</td>
<td>
2025-09-02
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>All clusters have an average probability of more than 75%, although
many outliers are observed. Finally, there are 829 (1.7 %) cells with an
assignment confidence of less than 75 %, showing that most of the cells
have been clustered very confidently.</p>
</div>
<div id="flightpath-exploration" class="section level2">
<h2>Flightpath exploration</h2>
<p>The InSituType package includes an interesting visualization tool
called “flightpath_plot”, which represents cells according to their
cluster probabilities. This will allow refining the clustering
assignation.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="do">## Code adapted from the InSituType vignette</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="co"># Create custom color palette</span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">gg_color_hue</span>(n_clus)</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="fu">names</span>(cols) <span class="ot">&lt;-</span> <span class="fu">unique</span>(unsup<span class="sc">$</span>clust)</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a></span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a><span class="co"># Flightpath plot</span></span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a><span class="fu">flightpath_plot</span>(<span class="at">insitutype_result =</span> unsup,</span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a>                <span class="at">col =</span> cols[unsup<span class="sc">$</span>clust])</span></code></pre></div>
<p><img src="figure/4.1_insitutype_unsup_clustering.Rmd/VizFPUnsup-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-VizFPUnsup-1">
Past versions of VizFPUnsup-1.png
</button>
</p>
<div id="fig-VizFPUnsup-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/lga-uoc/CosMx_pipeline_LGA/blob/43970ac7eece0aded282c805767d439617b6a2a0/docs/figure/4.1_insitutype_unsup_clustering.Rmd/VizFPUnsup-1.png" target="_blank">43970ac</a>
</td>
<td>
lga-uoc
</td>
<td>
2025-09-02
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>As before, it can be seen that all clusters are very well defined,
with average probabilities of nearly 100%. A few connections are
observed between some clusters, representing cells whose expression
profiles could be confused between said clusters.</p>
</div>
<div id="clustering-refining" class="section level2">
<h2>Clustering refining</h2>
<p>After exploration, using the function “refineClusters”, the user can
determine whether some clusters should be deleted, merged or
subclustered.</p>
<p>However, in this case, all clusters seem very well defined and
present good quality assignments scores, so no refining is needed.</p>
</div>
<div id="add-clustering-to-seurat-object" class="section level2">
<h2>Add clustering to Seurat object</h2>
<p>Now, the clusters can be included in the Seurat object for UMAP
visualization.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co"># Add new metadata column to seurat object</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>seu<span class="sc">$</span>IST_unsup_clus <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(unsup<span class="sc">$</span>clust)</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="co">#seu$IST_unsup_clus_refined &lt;- as.factor(newclusts$clust) # Refined clusters</span></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="co"># Set as default idents</span></span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a><span class="fu">Idents</span>(seu) <span class="ot">&lt;-</span> seu<span class="sc">$</span>IST_unsup_clus</span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="co"># Create custom color palette</span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">unique</span>(<span class="fu">Idents</span>(seu)))</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">gg_color_hue</span>(n)</span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="co"># DimPlot of selected resolution</span></span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a><span class="fu">DimPlot</span>(seu, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">cols =</span> cols, <span class="at">raster =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="figure/4.1_insitutype_unsup_clustering.Rmd/VizClus-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-VizClus-1">
Past versions of VizClus-1.png
</button>
</p>
<div id="fig-VizClus-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/lga-uoc/CosMx_pipeline_LGA/blob/43970ac7eece0aded282c805767d439617b6a2a0/docs/figure/4.1_insitutype_unsup_clustering.Rmd/VizClus-1.png" target="_blank">43970ac</a>
</td>
<td>
lga-uoc
</td>
<td>
2025-09-02
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Finally, the visualization of “nFeature” and “nCount” allows to
identify clusters with better/lower representation by the CosMx
panel.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="do">## Code adapted from CosMxLite vignette</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="do">## Points are removed as CosMx datasets can contain over 1 million cells and may hide the violin plot</span></span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a><span class="co"># Violin plot</span></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a><span class="fu">VlnPlot</span>(seu, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">&quot;nFeature_RNA&quot;</span>, <span class="st">&quot;nCount_RNA&quot;</span>), <span class="at">ncol =</span> <span class="dv">1</span>, <span class="at">pt.size =</span> <span class="dv">0</span>, <span class="at">group.by =</span> <span class="st">&quot;IST_unsup_clus&quot;</span>)</span></code></pre></div>
<p><img src="figure/4.1_insitutype_unsup_clustering.Rmd/VizClus2-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-VizClus2-1">
Past versions of VizClus2-1.png
</button>
</p>
<div id="fig-VizClus2-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/lga-uoc/CosMx_pipeline_LGA/blob/43970ac7eece0aded282c805767d439617b6a2a0/docs/figure/4.1_insitutype_unsup_clustering.Rmd/VizClus2-1.png" target="_blank">43970ac</a>
</td>
<td>
lga-uoc
</td>
<td>
2025-09-02
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="cell-annotation---sctype" class="section level1">
<h1>Cell annotation - scType</h1>
<p>In this script, clusters will be annotated based on their top markers
using the “scType” package and its “Brain” reference. For more
information about this method check its documentation and vignette <a
href="https://github.com/IanevskiAleksandr/sc-type">here</a>.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="do">## Code adapted from scType vignette</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a><span class="co"># Load gene set preparation and cell type annotation function</span></span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/gene_sets_prepare.R&quot;</span>)</span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/sctype_score_.R&quot;</span>)</span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a><span class="co"># Prepare DB file and select tissue</span></span>
<span id="cb17-8"><a href="#cb17-8" tabindex="-1"></a>db_ <span class="ot">&lt;-</span> <span class="st">&quot;https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/ScTypeDB_full.xlsx&quot;</span>;</span>
<span id="cb17-9"><a href="#cb17-9" tabindex="-1"></a>tissue <span class="ot">&lt;-</span> <span class="st">&quot;Brain&quot;</span> <span class="co"># e.g. Immune system, Pancreas, Liver, Kidney, Brain...</span></span>
<span id="cb17-10"><a href="#cb17-10" tabindex="-1"></a></span>
<span id="cb17-11"><a href="#cb17-11" tabindex="-1"></a><span class="co"># Prepare gene sets</span></span>
<span id="cb17-12"><a href="#cb17-12" tabindex="-1"></a>gs_list <span class="ot">&lt;-</span> <span class="fu">gene_sets_prepare</span>(db_, tissue)</span>
<span id="cb17-13"><a href="#cb17-13" tabindex="-1"></a></span>
<span id="cb17-14"><a href="#cb17-14" tabindex="-1"></a><span class="co"># Extract scaled data from the Seurat object</span></span>
<span id="cb17-15"><a href="#cb17-15" tabindex="-1"></a><span class="cf">if</span> (folder <span class="sc">==</span> <span class="st">&quot;SCT&quot;</span>) {</span>
<span id="cb17-16"><a href="#cb17-16" tabindex="-1"></a>  slot <span class="ot">&lt;-</span> <span class="st">&quot;SCT&quot;</span></span>
<span id="cb17-17"><a href="#cb17-17" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb17-18"><a href="#cb17-18" tabindex="-1"></a>  slot <span class="ot">&lt;-</span> <span class="st">&quot;RNA&quot;</span>  </span>
<span id="cb17-19"><a href="#cb17-19" tabindex="-1"></a>}</span>
<span id="cb17-20"><a href="#cb17-20" tabindex="-1"></a></span>
<span id="cb17-21"><a href="#cb17-21" tabindex="-1"></a>scaled_data <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(seu[[<span class="fu">paste0</span>(slot)]]<span class="sc">@</span>scale.data) </span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="do">## Code adapted from scType vignette</span></span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a><span class="co"># Run ScType</span></span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a>es.max <span class="ot">&lt;-</span> <span class="fu">sctype_score</span>(scaled_data, <span class="at">scaled =</span> <span class="cn">TRUE</span>, </span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a>                       <span class="at">gs =</span> gs_list<span class="sc">$</span>gs_positive, </span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a>                       <span class="at">gs2 =</span> gs_list<span class="sc">$</span>gs_negative)</span>
<span id="cb18-7"><a href="#cb18-7" tabindex="-1"></a></span>
<span id="cb18-8"><a href="#cb18-8" tabindex="-1"></a><span class="co"># Merge by cluster</span></span>
<span id="cb18-9"><a href="#cb18-9" tabindex="-1"></a>cL_resutls <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">&quot;rbind&quot;</span>, <span class="fu">lapply</span>(<span class="fu">unique</span>(seu<span class="sc">@</span>meta.data<span class="sc">$</span>IST_unsup_clus), <span class="cf">function</span>(cl) {</span>
<span id="cb18-10"><a href="#cb18-10" tabindex="-1"></a>    es.max.cl <span class="ot">=</span> <span class="fu">sort</span>(<span class="fu">rowSums</span>(es.max[ ,<span class="fu">rownames</span>(seu<span class="sc">@</span>meta.data[seu<span class="sc">@</span>meta.data<span class="sc">$</span>IST_unsup_clus<span class="sc">==</span>cl, ])]), </span>
<span id="cb18-11"><a href="#cb18-11" tabindex="-1"></a>                     <span class="at">decreasing =</span> <span class="sc">!</span><span class="dv">0</span>)</span>
<span id="cb18-12"><a href="#cb18-12" tabindex="-1"></a>    <span class="fu">head</span>(<span class="fu">data.frame</span>(<span class="at">cluster =</span> cl, <span class="at">type =</span> <span class="fu">names</span>(es.max.cl), <span class="at">scores =</span> es.max.cl, </span>
<span id="cb18-13"><a href="#cb18-13" tabindex="-1"></a>                    <span class="at">ncells =</span> <span class="fu">sum</span>(seu<span class="sc">@</span>meta.data<span class="sc">$</span>IST_unsup_clus<span class="sc">==</span>cl)), <span class="dv">10</span>)</span>
<span id="cb18-14"><a href="#cb18-14" tabindex="-1"></a>}))</span>
<span id="cb18-15"><a href="#cb18-15" tabindex="-1"></a></span>
<span id="cb18-16"><a href="#cb18-16" tabindex="-1"></a>sctype_scores <span class="ot">&lt;-</span> cL_resutls <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span> <span class="fu">top_n</span>(<span class="at">n =</span> <span class="dv">1</span>, <span class="at">wt =</span> scores)</span>
<span id="cb18-17"><a href="#cb18-17" tabindex="-1"></a></span>
<span id="cb18-18"><a href="#cb18-18" tabindex="-1"></a><span class="co"># Set low-confident (low ScType score) clusters to &quot;unknown&quot;</span></span>
<span id="cb18-19"><a href="#cb18-19" tabindex="-1"></a>sctype_scores<span class="sc">$</span>type[<span class="fu">as.numeric</span>(<span class="fu">as.character</span>(sctype_scores<span class="sc">$</span>scores)) <span class="sc">&lt;</span> sctype_scores<span class="sc">$</span>ncells<span class="sc">/</span><span class="dv">4</span>] <span class="ot">&lt;-</span> <span class="st">&quot;Unknown&quot;</span></span>
<span id="cb18-20"><a href="#cb18-20" tabindex="-1"></a></span>
<span id="cb18-21"><a href="#cb18-21" tabindex="-1"></a><span class="do">## Self coded addition</span></span>
<span id="cb18-22"><a href="#cb18-22" tabindex="-1"></a><span class="co"># Maintain original cluster separation</span></span>
<span id="cb18-23"><a href="#cb18-23" tabindex="-1"></a>sctype_scores<span class="sc">$</span>cl_type <span class="ot">&lt;-</span> <span class="fu">paste0</span>(sctype_scores<span class="sc">$</span>cluster, <span class="st">&quot;_&quot;</span>, sctype_scores<span class="sc">$</span>type)</span>
<span id="cb18-24"><a href="#cb18-24" tabindex="-1"></a></span>
<span id="cb18-25"><a href="#cb18-25" tabindex="-1"></a><span class="co"># Sort by cluster</span></span>
<span id="cb18-26"><a href="#cb18-26" tabindex="-1"></a>sctype_scores <span class="ot">&lt;-</span> sctype_scores <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(<span class="fu">as.numeric</span>(sctype_scores<span class="sc">$</span>cluster))</span>
<span id="cb18-27"><a href="#cb18-27" tabindex="-1"></a><span class="fu">print</span>(sctype_scores)</span></code></pre></div>
<pre><code># A tibble: 17 × 5
# Groups:   cluster [17]
   cluster type                             scores ncells cl_type               
   &lt;fct&gt;   &lt;chr&gt;                             &lt;dbl&gt;  &lt;int&gt; &lt;chr&gt;                 
 1 a       GABAergic neurons               16076.    2108 a_GABAergic neurons   
 2 b       Cancer stem cells                5091.    2908 b_Cancer stem cells   
 3 c       GABAergic neurons                1592.     154 c_GABAergic neurons   
 4 d       Endothelial cells               20564.    2805 d_Endothelial cells   
 5 e       Cholinergic neurons               324.      45 e_Cholinergic neurons 
 6 f       Microglial cells                 1605.     777 f_Microglial cells    
 7 g       Glutamatergic neurons            9202.   11800 g_Glutamatergic neuro…
 8 h       GABAergic neurons                  14.0      5 h_GABAergic neurons   
 9 i       Oligodendrocyte precursor cells  2800.    1393 i_Oligodendrocyte pre…
10 j       Oligodendrocytes                 4849.    1367 j_Oligodendrocytes    
11 k       Dopaminergic neurons             1054.     363 k_Dopaminergic neurons
12 l       Microglial cells                11669.    2214 l_Microglial cells    
13 m       GABAergic neurons               12823.     960 m_GABAergic neurons   
14 n       Glutamatergic neurons            8152.    4988 n_Glutamatergic neuro…
15 o       Oligodendrocytes                80506.    6378 o_Oligodendrocytes    
16 p       GABAergic neurons                6895.    3369 p_GABAergic neurons   
17 q       Astrocytes                      36276.    6284 q_Astrocytes          </code></pre>
<p>In this case the “Brain” database has confidently annotated all
clusters but has merged a lot clusters that got the same label: clusters
a, b, d, e and j as GABAergic neurons; c and h as Dopaminergic neurons;
g and q as Glutamatergic neurons; f and i as Oligodendrocytes; and o and
p as Cancer stem cells.</p>
<div id="add-annotations-to-seurat-object" class="section level2">
<h2>Add annotations to Seurat object</h2>
<p>In order not to loose information, both the scType annotation and a
“cluster+scType” annotation will be saved in the seurat object.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="co"># Convert to data.frame and select columns</span></span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a>annot_df <span class="ot">&lt;-</span> sctype_scores <span class="sc">%&gt;%</span> </span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(cluster, type, cl_type) <span class="sc">%&gt;%</span></span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">cluster =</span> <span class="fu">as.character</span>(cluster))</span>
<span id="cb20-5"><a href="#cb20-5" tabindex="-1"></a></span>
<span id="cb20-6"><a href="#cb20-6" tabindex="-1"></a><span class="co"># Create named vectors for easy mapping</span></span>
<span id="cb20-7"><a href="#cb20-7" tabindex="-1"></a>scType_ann <span class="ot">&lt;-</span> <span class="fu">setNames</span>(annot_df<span class="sc">$</span>type, annot_df<span class="sc">$</span>cluster)</span>
<span id="cb20-8"><a href="#cb20-8" tabindex="-1"></a>cl_scType_ann <span class="ot">&lt;-</span> <span class="fu">setNames</span>(annot_df<span class="sc">$</span>cl_type, annot_df<span class="sc">$</span>cluster)</span>
<span id="cb20-9"><a href="#cb20-9" tabindex="-1"></a></span>
<span id="cb20-10"><a href="#cb20-10" tabindex="-1"></a><span class="co"># Add new metadata columns to seurat object</span></span>
<span id="cb20-11"><a href="#cb20-11" tabindex="-1"></a>seu<span class="sc">$</span>scType_ann <span class="ot">&lt;-</span> <span class="fu">factor</span>(scType_ann[<span class="fu">as.character</span>(seu<span class="sc">$</span>IST_unsup_clus)],</span>
<span id="cb20-12"><a href="#cb20-12" tabindex="-1"></a>                         <span class="at">levels =</span> <span class="fu">unique</span>(annot_df<span class="sc">$</span>type))</span>
<span id="cb20-13"><a href="#cb20-13" tabindex="-1"></a></span>
<span id="cb20-14"><a href="#cb20-14" tabindex="-1"></a>seu<span class="sc">$</span>cl_scType_ann <span class="ot">&lt;-</span> <span class="fu">factor</span>(cl_scType_ann[<span class="fu">as.character</span>(seu<span class="sc">$</span>IST_unsup_clus)],</span>
<span id="cb20-15"><a href="#cb20-15" tabindex="-1"></a>                            <span class="at">levels =</span> annot_df<span class="sc">$</span>cl_type)</span>
<span id="cb20-16"><a href="#cb20-16" tabindex="-1"></a></span>
<span id="cb20-17"><a href="#cb20-17" tabindex="-1"></a><span class="co"># Set the cl_scType annotation as default</span></span>
<span id="cb20-18"><a href="#cb20-18" tabindex="-1"></a><span class="fu">Idents</span>(seu) <span class="ot">&lt;-</span> seu<span class="sc">$</span>scType_ann</span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="co"># Save old custom palette for original clusters</span></span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>n_cl <span class="ot">&lt;-</span> n</span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a>cols_cl <span class="ot">&lt;-</span> cols</span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a><span class="co"># Create custom color palettes</span></span>
<span id="cb21-6"><a href="#cb21-6" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">unique</span>(<span class="fu">Idents</span>(seu)))</span>
<span id="cb21-7"><a href="#cb21-7" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">gg_color_hue</span>(n)</span>
<span id="cb21-8"><a href="#cb21-8" tabindex="-1"></a></span>
<span id="cb21-9"><a href="#cb21-9" tabindex="-1"></a><span class="co"># Inspect UMAP scType annotation</span></span>
<span id="cb21-10"><a href="#cb21-10" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seu, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="at">label =</span> <span class="cn">FALSE</span>, <span class="at">raster =</span> <span class="cn">FALSE</span>, </span>
<span id="cb21-11"><a href="#cb21-11" tabindex="-1"></a>              <span class="at">repel =</span> <span class="cn">TRUE</span>, <span class="at">cols =</span> cols_cl, <span class="at">group.by =</span> <span class="st">&quot;cl_scType_ann&quot;</span>)</span>
<span id="cb21-12"><a href="#cb21-12" tabindex="-1"></a></span>
<span id="cb21-13"><a href="#cb21-13" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seu, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="at">label =</span> <span class="cn">FALSE</span>, <span class="at">raster =</span> <span class="cn">FALSE</span>, </span>
<span id="cb21-14"><a href="#cb21-14" tabindex="-1"></a>              <span class="at">repel =</span> <span class="cn">TRUE</span>, <span class="at">cols =</span> cols, <span class="at">group.by =</span> <span class="st">&quot;scType_ann&quot;</span>)</span>
<span id="cb21-15"><a href="#cb21-15" tabindex="-1"></a></span>
<span id="cb21-16"><a href="#cb21-16" tabindex="-1"></a><span class="co"># Arrange plots</span></span>
<span id="cb21-17"><a href="#cb21-17" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code></pre></div>
<p><img src="figure/4.1_insitutype_unsup_clustering.Rmd/VizAnn-1.png" width="1920" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-VizAnn-1">
Past versions of VizAnn-1.png
</button>
</p>
<div id="fig-VizAnn-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/lga-uoc/CosMx_pipeline_LGA/blob/43970ac7eece0aded282c805767d439617b6a2a0/docs/figure/4.1_insitutype_unsup_clustering.Rmd/VizAnn-1.png" target="_blank">43970ac</a>
</td>
<td>
lga-uoc
</td>
<td>
2025-09-02
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="save-the-seurat-object" class="section level2">
<h2>Save the Seurat object</h2>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="co"># Save the annotated Seurat object</span></span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>name <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;seu_&quot;</span>, folder, <span class="st">&quot;_ISTUnsup.RDS&quot;</span>)</span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a><span class="fu">saveRDS</span>(seu, <span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;processed_data&quot;</span>,folder,name))</span></code></pre></div>
</div>
</div>
<div id="final-spatial-visualizations" class="section level1">
<h1>Final spatial visualizations</h1>
<p>Now that cells have been annotated, visualizing them in their spatial
context becomes more informative. For example, all cells can be
visualized simultaneously in the sample or one cell type can be
specifically highlighted:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="do">## Code adapted from Seurat Spatial vignette</span></span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ImageDimPlot</span>(seu, <span class="at">fov =</span> <span class="st">&quot;globalFOV&quot;</span>, <span class="at">axes =</span> <span class="cn">TRUE</span>, <span class="at">cols =</span> cols) <span class="sc">+</span></span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a></span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ImageDimPlot</span>(seu, <span class="at">fov =</span> <span class="st">&quot;globalFOV&quot;</span>, <span class="at">axes =</span> <span class="cn">TRUE</span>, <span class="at">cols =</span> cols[<span class="dv">3</span>],</span>
<span id="cb23-6"><a href="#cb23-6" tabindex="-1"></a>                   <span class="at">cells =</span> <span class="fu">WhichCells</span>(seu, <span class="at">idents =</span> <span class="st">&quot;Oligodendrocytes&quot;</span>)) <span class="sc">+</span></span>
<span id="cb23-7"><a href="#cb23-7" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Oligodendrocytes&quot;</span>) <span class="sc">+</span> </span>
<span id="cb23-8"><a href="#cb23-8" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb23-9"><a href="#cb23-9" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">face =</span> <span class="st">&quot;bold&quot;</span>, <span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb23-10"><a href="#cb23-10" tabindex="-1"></a></span>
<span id="cb23-11"><a href="#cb23-11" tabindex="-1"></a><span class="co"># Arrange plots</span></span>
<span id="cb23-12"><a href="#cb23-12" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code></pre></div>
<p><img src="figure/4.1_insitutype_unsup_clustering.Rmd/SpatialViz-1.png" width="3900" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-SpatialViz-1">
Past versions of SpatialViz-1.png
</button>
</p>
<div id="fig-SpatialViz-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/lga-uoc/CosMx_pipeline_LGA/blob/43970ac7eece0aded282c805767d439617b6a2a0/docs/figure/4.1_insitutype_unsup_clustering.Rmd/SpatialViz-1.png" target="_blank">43970ac</a>
</td>
<td>
lga-uoc
</td>
<td>
2025-09-02
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Another interesting visualization would be to plot gene expression
markers in its spatial context. For example, one typical maker for
Oligodendrocytes is <em>Olig1</em>:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="do">## Code adapted from Seurat Spatial vignette</span></span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a><span class="do">## Points are removed from violin plots as CosMx datasets can contain over 1 million cells and may hide the violin plot</span></span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" tabindex="-1"></a><span class="fu">VlnPlot</span>(seu, <span class="at">features =</span> <span class="st">&quot;Olig1&quot;</span>, <span class="at">pt.size =</span> <span class="dv">0</span>, <span class="at">y.max =</span> <span class="dv">20</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code></pre></div>
<p><img src="figure/4.1_insitutype_unsup_clustering.Rmd/MarkerViz1-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-MarkerViz1-1">
Past versions of MarkerViz1-1.png
</button>
</p>
<div id="fig-MarkerViz1-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/lga-uoc/CosMx_pipeline_LGA/blob/43970ac7eece0aded282c805767d439617b6a2a0/docs/figure/4.1_insitutype_unsup_clustering.Rmd/MarkerViz1-1.png" target="_blank">43970ac</a>
</td>
<td>
lga-uoc
</td>
<td>
2025-09-02
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>As expected, it is highly expressed in Oligodendrocytes and
Oligodendrocyte precursor cells, but it can also be observed in its
spatial context:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="do">## Code adapted from CosMxLite vignette</span></span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ImageFeaturePlot</span>(seu, <span class="at">fov =</span> <span class="st">&quot;globalFOV&quot;</span>, <span class="at">features =</span> <span class="st">&quot;Olig1&quot;</span>) <span class="sc">+</span></span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb25-4"><a href="#cb25-4" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">face =</span> <span class="st">&quot;bold&quot;</span>, <span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb25-5"><a href="#cb25-5" tabindex="-1"></a></span>
<span id="cb25-6"><a href="#cb25-6" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ImageDimPlot</span>(seu, <span class="at">fov =</span> <span class="st">&quot;globalFOV&quot;</span>, <span class="at">axes =</span> <span class="cn">TRUE</span>, <span class="at">cols =</span> cols[<span class="dv">3</span>],</span>
<span id="cb25-7"><a href="#cb25-7" tabindex="-1"></a>                   <span class="at">cells =</span> <span class="fu">WhichCells</span>(seu, <span class="at">idents =</span> <span class="st">&quot;Oligodendrocytes&quot;</span>)) <span class="sc">+</span></span>
<span id="cb25-8"><a href="#cb25-8" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Oligodendrocytes&quot;</span>) <span class="sc">+</span> </span>
<span id="cb25-9"><a href="#cb25-9" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb25-10"><a href="#cb25-10" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">face =</span> <span class="st">&quot;bold&quot;</span>, <span class="at">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb25-11"><a href="#cb25-11" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">&quot;bottom&quot;</span>)</span>
<span id="cb25-12"><a href="#cb25-12" tabindex="-1"></a></span>
<span id="cb25-13"><a href="#cb25-13" tabindex="-1"></a><span class="co"># Arrange plots</span></span>
<span id="cb25-14"><a href="#cb25-14" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code></pre></div>
<p><img src="figure/4.1_insitutype_unsup_clustering.Rmd/SpatialViz2-1.png" width="3900" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-SpatialViz2-1">
Past versions of SpatialViz2-1.png
</button>
</p>
<div id="fig-SpatialViz2-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/lga-uoc/CosMx_pipeline_LGA/blob/43970ac7eece0aded282c805767d439617b6a2a0/docs/figure/4.1_insitutype_unsup_clustering.Rmd/SpatialViz2-1.png" target="_blank">43970ac</a>
</td>
<td>
lga-uoc
</td>
<td>
2025-09-02
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="performance-and-session-info" class="section level1">
<h1>Performance and Session Info</h1>
<details>
<summary>
<strong>Performance Report</strong>
</summary>
<table class="table table-striped table-hover table-condensed" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
Chunk
</th>
<th style="text-align:right;">
Time_sec
</th>
<th style="text-align:right;">
Memory_Mb
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Libraries
</td>
<td style="text-align:right;">
4.22
</td>
<td style="text-align:right;">
459.1
</td>
</tr>
<tr>
<td style="text-align:left;">
LoadData
</td>
<td style="text-align:right;">
5.04
</td>
<td style="text-align:right;">
1215.0
</td>
</tr>
<tr>
<td style="text-align:left;">
PrepData
</td>
<td style="text-align:right;">
0.37
</td>
<td style="text-align:right;">
205.5
</td>
</tr>
<tr>
<td style="text-align:left;">
PrepCohort
</td>
<td style="text-align:right;">
7.02
</td>
<td style="text-align:right;">
2.6
</td>
</tr>
<tr>
<td style="text-align:left;">
Unsup
</td>
<td style="text-align:right;">
60.40
</td>
<td style="text-align:right;">
16.6
</td>
</tr>
<tr>
<td style="text-align:left;">
UnsupProb
</td>
<td style="text-align:right;">
0.30
</td>
<td style="text-align:right;">
0.1
</td>
</tr>
<tr>
<td style="text-align:left;">
ExporingUnsupProb
</td>
<td style="text-align:right;">
0.29
</td>
<td style="text-align:right;">
0.0
</td>
</tr>
<tr>
<td style="text-align:left;">
VizUnsupProb
</td>
<td style="text-align:right;">
1.01
</td>
<td style="text-align:right;">
14.1
</td>
</tr>
<tr>
<td style="text-align:left;">
VizFPUnsup
</td>
<td style="text-align:right;">
3.07
</td>
<td style="text-align:right;">
12.9
</td>
</tr>
<tr>
<td style="text-align:left;">
RefinedUnsup
</td>
<td style="text-align:right;">
0.31
</td>
<td style="text-align:right;">
0.1
</td>
</tr>
<tr>
<td style="text-align:left;">
AddClus
</td>
<td style="text-align:right;">
0.32
</td>
<td style="text-align:right;">
0.5
</td>
</tr>
<tr>
<td style="text-align:left;">
VizClus
</td>
<td style="text-align:right;">
1.62
</td>
<td style="text-align:right;">
8.9
</td>
</tr>
<tr>
<td style="text-align:left;">
VizClus2
</td>
<td style="text-align:right;">
0.89
</td>
<td style="text-align:right;">
2.4
</td>
</tr>
<tr>
<td style="text-align:left;">
scTypePrep
</td>
<td style="text-align:right;">
1.47
</td>
<td style="text-align:right;">
16.4
</td>
</tr>
<tr>
<td style="text-align:left;">
RunscType
</td>
<td style="text-align:right;">
4.67
</td>
<td style="text-align:right;">
14.7
</td>
</tr>
<tr>
<td style="text-align:left;">
AddAnn
</td>
<td style="text-align:right;">
0.37
</td>
<td style="text-align:right;">
1.1
</td>
</tr>
<tr>
<td style="text-align:left;">
VizAnn
</td>
<td style="text-align:right;">
2.95
</td>
<td style="text-align:right;">
13.0
</td>
</tr>
<tr>
<td style="text-align:left;">
SavingSeuObj
</td>
<td style="text-align:right;">
34.07
</td>
<td style="text-align:right;">
0.0
</td>
</tr>
<tr>
<td style="text-align:left;">
SpatialViz
</td>
<td style="text-align:right;">
3.55
</td>
<td style="text-align:right;">
11.8
</td>
</tr>
<tr>
<td style="text-align:left;">
MarkerViz1
</td>
<td style="text-align:right;">
0.61
</td>
<td style="text-align:right;">
2.9
</td>
</tr>
<tr>
<td style="text-align:left;">
SpatialViz2
</td>
<td style="text-align:right;">
3.60
</td>
<td style="text-align:right;">
8.7
</td>
</tr>
<tr>
<td style="text-align:left;">
Total
</td>
<td style="text-align:right;">
136.15
</td>
<td style="text-align:right;">
2006.4
</td>
</tr>
</tbody>
</table>
</details>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>R version 4.4.3 (2025-02-28 ucrt)
Platform: x86_64-w64-mingw32/x64
Running under: Windows 10 x64 (build 19045)

Matrix products: default


locale:
[1] LC_COLLATE=Spanish_Spain.utf8  LC_CTYPE=Spanish_Spain.utf8   
[3] LC_MONETARY=Spanish_Spain.utf8 LC_NUMERIC=C                  
[5] LC_TIME=Spanish_Spain.utf8    

time zone: Europe/Madrid
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] InSituType_2.0     HGNChelper_0.8.15  pheatmap_1.0.12    patchwork_1.3.0   
 [5] ggplot2_3.5.1      SeuratObject_4.1.4 Seurat_4.4.0       kableExtra_1.4.0  
 [9] dplyr_1.1.4        here_1.0.1         Matrix_1.7-2       data.table_1.17.0 
[13] workflowr_1.7.1   

loaded via a namespace (and not attached):
  [1] RcppAnnoy_0.0.22            splines_4.4.3              
  [3] later_1.4.1                 tibble_3.2.1               
  [5] polyclip_1.10-7             lifecycle_1.0.4            
  [7] rprojroot_2.0.4             globals_0.16.3             
  [9] processx_3.8.4              lattice_0.22-6             
 [11] MASS_7.3-64                 SnowballC_0.7.1            
 [13] magrittr_2.0.3              openxlsx_4.2.8             
 [15] plotly_4.11.0               sass_0.4.9                 
 [17] rmarkdown_2.29              jquerylib_0.1.4            
 [19] yaml_2.3.10                 httpuv_1.6.15              
 [21] sctransform_0.4.2           zip_2.3.1                  
 [23] askpass_1.2.1               sp_2.2-0                   
 [25] spatstat.sparse_3.1-0       reticulate_1.40.0          
 [27] cowplot_1.1.3               pbapply_1.7-4              
 [29] RColorBrewer_1.1-3          abind_1.4-8                
 [31] zlibbioc_1.52.0             Rtsne_0.17                 
 [33] GenomicRanges_1.58.0        purrr_1.0.2                
 [35] BiocGenerics_0.52.0         git2r_0.36.2               
 [37] GenomeInfoDbData_1.2.13     IRanges_2.40.1             
 [39] S4Vectors_0.44.0            ggrepel_0.9.6              
 [41] irlba_2.3.5.1               listenv_0.9.1              
 [43] spatstat.utils_3.1-5        umap_0.2.10.0              
 [45] goftest_1.2-3               RSpectra_0.16-2            
 [47] spatstat.random_3.4-1       fitdistrplus_1.2-4         
 [49] parallelly_1.41.0           svglite_2.1.3              
 [51] leiden_0.4.3.1              codetools_0.2-20           
 [53] DelayedArray_0.32.0         xml2_1.3.6                 
 [55] tidyselect_1.2.1            UCSC.utils_1.2.0           
 [57] farver_2.1.2                matrixStats_1.4.1          
 [59] stats4_4.4.3                spatstat.explore_3.5-2     
 [61] jsonlite_1.8.9              progressr_0.15.1           
 [63] ggridges_0.5.6              survival_3.8-3             
 [65] systemfonts_1.1.0           tools_4.4.3                
 [67] ragg_1.3.3                  ica_1.0-3                  
 [69] Rcpp_1.0.13-1               glue_1.8.0                 
 [71] gridExtra_2.3               SparseArray_1.6.0          
 [73] xfun_0.49                   MatrixGenerics_1.18.0      
 [75] GenomeInfoDb_1.42.1         withr_3.0.2                
 [77] fastmap_1.2.0               openssl_2.2.2              
 [79] callr_3.7.6                 digest_0.6.37              
 [81] R6_2.5.1                    mime_0.12                  
 [83] textshaping_0.4.1           colorspace_2.1-1           
 [85] scattermore_1.2             tensor_1.5.1               
 [87] spatstat.data_3.1-6         utf8_1.2.4                 
 [89] tidyr_1.3.1                 generics_0.1.3             
 [91] httr_1.4.7                  htmlwidgets_1.6.4          
 [93] S4Arrays_1.6.0              whisker_0.4.1              
 [95] uwot_0.2.3                  pkgconfig_2.0.3            
 [97] gtable_0.3.6                lmtest_0.9-40              
 [99] SingleCellExperiment_1.28.1 XVector_0.46.0             
[101] htmltools_0.5.8.1           scales_1.3.0               
[103] Biobase_2.66.0              png_0.1-8                  
[105] spatstat.univar_3.1-4       knitr_1.49                 
[107] rstudioapi_0.17.1           reshape2_1.4.4             
[109] nlme_3.1-167                cachem_1.1.0               
[111] zoo_1.8-14                  stringr_1.5.1              
[113] KernSmooth_2.23-26          parallel_4.4.3             
[115] miniUI_0.1.1.1              pillar_1.10.0              
[117] grid_4.4.3                  vctrs_0.6.5                
[119] RANN_2.6.2                  lsa_0.73.3                 
[121] promises_1.3.2              xtable_1.8-4               
[123] cluster_2.1.8               evaluate_1.0.1             
[125] cli_3.6.3                   compiler_4.4.3             
[127] rlang_1.1.4                 crayon_1.5.3               
[129] future.apply_1.11.3         labeling_0.4.3             
[131] mclust_6.1.1                ps_1.8.1                   
[133] getPass_0.2-4               plyr_1.8.9                 
[135] fs_1.6.5                    stringi_1.8.4              
[137] viridisLite_0.4.2           deldir_2.0-4               
[139] munsell_0.5.1               lazyeval_0.2.2             
[141] spatstat.geom_3.5-0         sparseMatrixStats_1.18.0   
[143] future_1.34.0               shiny_1.10.0               
[145] SummarizedExperiment_1.36.0 ROCR_1.0-11                
[147] igraph_2.1.2                bslib_0.8.0                
[149] splitstackshape_1.4.8      </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
