<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Example - CosMx Data Dimensional reduction with Log normalized object</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #204a87; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #204a87; font-weight: bold; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #ce5c00; font-weight: bold; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>







<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">CosMx™ SMI Data Analysis Pipeline</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="dataset.html">Dataset</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pipeline
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="pipeline.html">Introduction</a>
    </li>
    <li>
      <a href="0.0_data_loading.html">Data Loading</a>
    </li>
    <li>
      <a href="1.0_qc_and_filtering.html">QC and filtering</a>
    </li>
    <li>
      <a href="2.0_normalization.html">Normalization</a>
    </li>
    <li>
      <a href="3.0_dimensional_reduction.html">Dimensional reduction</a>
    </li>
    <li>
      <a href="4.0_insitutype_cell_typing.html">Cell typing</a>
    </li>
  </ul>
</li>
<li>
  <a href="norm_examples.html">Normalization</a>
</li>
<li>
  <a href="clus_examples.html">Cell typing</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
<li>
  <a href="https://github.com/lidiaga/CosMx_pipeline_LGA">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Example - CosMx Data Dimensional reduction
with Log normalized object</h1>
<h4 class="author">LGA</h4>
<h4 class="date">02 septiembre 2025</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-09-02
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>CosMx_pipeline_LGA/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20250517code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20250517)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20250517code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20250517)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrong9bc387c">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong> 9bc387c
</a>
</p>
</div>
<div id="strongRepositoryversionstrong9bc387c"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version 9bc387c.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    NBClust-Plots/
    Ignored:    analysis/.Rhistory
    Ignored:    data/seuAtoMx/
    Ignored:    output/processed_data/Log/
    Ignored:    output/processed_data/RC/
    Ignored:    output/processed_data/SCT/
    Ignored:    output/processed_data/exprMat_unfiltered.RDS
    Ignored:    output/processed_data/fov_positions_unfiltered.RDS
    Ignored:    output/processed_data/metadata_unfiltered.RDS
    Ignored:    output/processed_data/negMat_unfiltered.RDS
    Ignored:    output/processed_data/seu_filtered.RDS
    Ignored:    output/processed_data/seu_semifiltered.RDS

Untracked files:
    Untracked:  output/performance_reports/

Unstaged changes:
    Modified:   CosMx_pipeline_LGA.Rproj
    Modified:   _workflowr.yml

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>analysis/6.1_Log_dimensional_reduction.Rmd</code>) and HTML
(<code>docs/6.1_Log_dimensional_reduction.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
78c8449
</td>
<td>
lga-uoc
</td>
<td>
2025-09-02
</td>
<td>
Add Rmds files in analysis
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<!-- This pipeline has been developed using Seurat v4.4, as Seurat v5 presents several changes to the object structure that affected the working methods. Aditionally, the AtoMx exported Seurat objects come in the Seurat v4 structure. Therefore, working in this version can be more compatible with exported data too. -->
<div id="dependencies" class="section level1">
<h1>Dependencies</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(data.table) <span class="co"># Efficient data management</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(here) <span class="co"># Enhanced file referencing in project-oriented workflows</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">library</span>(dplyr) <span class="co"># For the use of pipes %&gt;%</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">library</span>(kableExtra) <span class="co"># For table formatting</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="fu">library</span>(Seurat) <span class="co"># Seurat object</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="fu">library</span>(ggplot2) <span class="co"># Graphics</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="fu">library</span>(patchwork) <span class="co"># Layout graphics</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="fu">library</span>(viridis) <span class="co"># Viridis color scale</span></span></code></pre></div>
</div>
<div id="load-the-data" class="section level1">
<h1>Load the data</h1>
<p>First of all, data needs to be loaded into the session. For this
script, only the normalized/transformed Seurat object is needed.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Indicate the object folder</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>folder <span class="ot">&lt;-</span> <span class="st">&quot;Log&quot;</span> <span class="co"># Choose between &quot;RC&quot;, &quot;Log&quot; or &quot;SCT&quot;</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span>folder <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;RC&quot;</span>, <span class="st">&quot;Log&quot;</span>, <span class="st">&quot;SCT&quot;</span>)) {</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">&quot;The selected folder is invalid, choose: &#39;RC&#39;, &#39;Log&#39; or &#39;SCT&#39;&quot;</span>)</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>}</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co"># Load Seurat object</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>name <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;seu_&quot;</span>, folder, <span class="st">&quot;.RDS&quot;</span>)</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>seu <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;processed_data&quot;</span>,folder,name))</span></code></pre></div>
</div>
<div id="pca" class="section level1">
<h1>PCA</h1>
<p>Now that the data is normalized and scaled, the next step is to
perform linear dimensional reduction with Principal Components Analysis,
or PCA.</p>
<p>One of the benefits of using a Seurat object along the pipeline,
apart from keeping the data organised, is that well established and
extended packages, like Seurat, offer a great variety of functions that
cover the most important steps of a genomic analysis, specifically
designed to work with the object. Therefore, in this pipeline, PCA will
be executed using the Seurat function “RunPCA()”, followed by different
visualizations performed with different functions of this same
package.</p>
<!-- In this case, as the panel covers only 1K genes, all of them will be used in the PCA. However, if analyzing data from a 6K panel or a WTx panel. Only variable features can be selected setting *features* to VariableFeatures(seu). -->
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="do">## Code adapted from CosMxLite vignette</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="co"># Run PCA</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>seu <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(seu, <span class="at">features =</span> <span class="fu">rownames</span>(seu))</span></code></pre></div>
<p>One way of exploring the resulting PCs would be to visualize the
genes that are contributing the most to each PC, for example, by
printing the resulting list of genes with most positive and negative
loadings or using “VizDimLoadings()”.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="do">## Code adapted from CosMxLite vignette</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co"># Run PCA</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="fu">print</span>(seu[[<span class="st">&quot;pca&quot;</span>]], <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="at">nfeatures =</span> <span class="dv">5</span>)</span></code></pre></div>
<pre><code>PC_ 1 
Positive:  Apoe, Glul, Apod, Atp1a2, Ptgds 
Negative:  Dnm1, Syp, Ndrg4, Syt1, Snap25 
PC_ 2 
Positive:  Snap25, Atp1b1, Rtn1, Syp, Syt1 
Negative:  Daam2, Pllp, Mag, Apod, Cd9 
PC_ 3 
Positive:  Atp1a2, Gja1, Cst3, Clu, Gpr37l1 
Negative:  Mog, Mag, Plp1, Ugt8a, Fa2h 
PC_ 4 
Positive:  Nrgn, Cldn5, Flt1, Itm2a, Esam 
Negative:  Aldoc, Slc6a11, Ndrg2, Slc6a1, Nap1l5 
PC_ 5 
Positive:  Nrgn, Gpm6b, Malat1, Gpr37l1, Gria2 
Negative:  Bsg, Itm2a, Flt1, Sparc, Cldn5 </code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">VizDimLoadings</span>(seu, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, <span class="at">reduction =</span> <span class="st">&#39;pca&#39;</span>)</span></code></pre></div>
<p><img src="figure/6.1_Log_dimensional_reduction.Rmd/VizPCA1-1.png" width="576" style="display: block; margin: auto;" />
The loadings for each PC show which genes are driving the variance in
the data along that axis. In this case, for example, cells with a high
score for PC 1 would have a higher expression of genes like Apoe, Glul,
Apod, Atp1a2, Ptgds, while cells with a lower score for PC 1, would
instead have higher expression of genes like Dnm1, Syp, Ndrg4, Syt1,
Snap25.</p>
<p>Some other the visualization tools that allow for easy exploration of
the resulted PCs are “DimPlot()”, “DimHeatmap()” and “ElbowPlot()”:</p>
<div id="pca-dimplot" class="section level2">
<h2>PCA DimPlot</h2>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="do">## Code adapted from CosMxLite vignette</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="co"># DimPlot</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="fu">DimPlot</span>(seu, <span class="at">reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="at">raster =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;PCA DimPlot&quot;</span>) <span class="sc">+</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">face =</span> <span class="st">&quot;bold&quot;</span>)) <span class="sc">+</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>  <span class="fu">NoLegend</span>()</span></code></pre></div>
<p><img src="figure/6.1_Log_dimensional_reduction.Rmd/VizPCA2-1.png" width="384" style="display: block; margin: auto;" />
In this plot, cells that are far from each other in the X-axis (PC 1)
are more different than cells separated the same distance in the Y-axis
(PC 2). In this case, no clear differentiations are observed.</p>
</div>
<div id="pca-dimheatmap" class="section level2">
<h2>PCA DimHeatmap</h2>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="do">## Code adapted from CosMxLite vignette</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="co"># DimHeatmap</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>heat_list <span class="ot">&lt;-</span> <span class="fu">DimHeatmap</span>(seu, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, <span class="at">cells =</span> <span class="dv">500</span>, <span class="at">balanced =</span> <span class="cn">TRUE</span>,</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>                        <span class="at">fast =</span> <span class="cn">FALSE</span>, <span class="at">combine =</span> <span class="cn">TRUE</span>)</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co"># Add individual titles</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a>titles <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;PC&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>)</span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a>heat_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">seq_along</span>(heat_list), <span class="cf">function</span>(i) {</span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a>  heat_list[[i]] <span class="sc">+</span> <span class="fu">ggtitle</span>(titles[i])</span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a>})</span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="co"># Combine with global title</span></span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a><span class="fu">wrap_plots</span>(heat_list, <span class="at">guides =</span> <span class="st">&quot;collect&quot;</span>) <span class="sc">+</span> </span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a>  <span class="fu">plot_annotation</span>(</span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Multiple PCA DimHeatmaps&quot;</span>,</span>
<span id="cb8-18"><a href="#cb8-18" tabindex="-1"></a>    <span class="at">theme =</span> <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">face =</span> <span class="st">&quot;bold&quot;</span>))</span>
<span id="cb8-19"><a href="#cb8-19" tabindex="-1"></a>    ) <span class="sc">&amp;</span></span>
<span id="cb8-20"><a href="#cb8-20" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;right&quot;</span>)</span></code></pre></div>
<p><img src="figure/6.1_Log_dimensional_reduction.Rmd/VizPCA3-1.png" width="1440" style="display: block; margin: auto;" />
This type of plots represent cells (ranked through its PC score) in
columns and genes (top loadings) in rows, the color in each intersection
represents the expression level of the gene in the cell. By setting the
“cells” parameter to a number, such as 500, the plot only shows the most
extreme cells of each end. Therefor, the more clear the “blocks” of high
vs low expressing genes, the more likely that PC is capturing a major
biological distinction. In this case, for example, most PCs do not have
good differentiated blocks, instead they all look pretty fuzzy.</p>
</div>
<div id="elbow-plot" class="section level2">
<h2>Elbow Plot</h2>
<p>Finally, the Elbow Plot, or Scree Plot, ranks PCs based on how much
variance is explained by them. This type of plot is usually used to
determine the number of PCs that would be consider to “represent” the
dataset, leaving out those which could be considered to represent
technical noise.</p>
<p>Technically, the elbow method determines the optimal PC number in the
“elbow” of the curve. However, this might be subjective and there are
some quantitative approaches that can help determine the appropriate
number (see an example in the <a
href="https://hbctraining.github.io/scRNA-seq/lessons/elbow_plot_metric.html">Harvard
Chan Bioinformatics Core (HBC) scRNA-seq trainning</a>).</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="do">## Code adapted from hbctraining.github.io</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="co"># Metric 1 (first PC with &lt;5% variance but &gt;90% of cumulative variance)</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>pct <span class="ot">&lt;-</span> seu[[<span class="st">&quot;pca&quot;</span>]]<span class="sc">@</span>stdev <span class="sc">/</span> <span class="fu">sum</span>(seu[[<span class="st">&quot;pca&quot;</span>]]<span class="sc">@</span>stdev) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>cumu <span class="ot">&lt;-</span> <span class="fu">cumsum</span>(pct)</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>m1 <span class="ot">&lt;-</span> <span class="fu">which</span>(cumu <span class="sc">&gt;</span> <span class="dv">90</span> <span class="sc">&amp;</span> pct <span class="sc">&lt;</span> <span class="dv">5</span>)[<span class="dv">1</span>]</span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a><span class="co"># Metric 2 (first PC with &lt;0.1% variance difference with the consecutive PC)</span></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a>m2 <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">which</span>((pct[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(pct) <span class="sc">-</span> <span class="dv">1</span>] <span class="sc">-</span> pct[<span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(pct)]) <span class="sc">&gt;</span> <span class="fl">0.1</span>),</span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a>           <span class="at">decreasing =</span> T)[<span class="dv">1</span>] <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a></span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a></span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a><span class="do">## Code adapted from CosMxLite vignette</span></span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a></span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a><span class="co"># Elbow plot</span></span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a>npcs <span class="ot">&lt;-</span> <span class="dv">30</span> <span class="co"># Select the PCs to use</span></span>
<span id="cb9-18"><a href="#cb9-18" tabindex="-1"></a></span>
<span id="cb9-19"><a href="#cb9-19" tabindex="-1"></a><span class="fu">ElbowPlot</span>(seu, <span class="at">ndims =</span> <span class="dv">50</span>) <span class="sc">+</span> </span>
<span id="cb9-20"><a href="#cb9-20" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> m1, <span class="at">linetype =</span> <span class="st">&quot;dotted&quot;</span>, <span class="at">color =</span> <span class="st">&quot;blue&quot;</span>) <span class="sc">+</span> </span>
<span id="cb9-21"><a href="#cb9-21" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">&quot;text&quot;</span>, <span class="at">x =</span> m1, <span class="at">y =</span> <span class="dv">10</span>, <span class="at">label =</span> <span class="st">&quot;m1&quot;</span>, <span class="at">hjust =</span> <span class="sc">-</span><span class="fl">0.5</span>, <span class="at">color =</span> <span class="st">&quot;blue&quot;</span>) <span class="sc">+</span></span>
<span id="cb9-22"><a href="#cb9-22" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> m2, <span class="at">linetype =</span> <span class="st">&quot;dotted&quot;</span>, <span class="at">color =</span> <span class="st">&quot;blue&quot;</span>) <span class="sc">+</span> </span>
<span id="cb9-23"><a href="#cb9-23" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">&quot;text&quot;</span>, <span class="at">x =</span> m2, <span class="at">y =</span> <span class="dv">10</span>, <span class="at">label =</span> <span class="st">&quot;m2&quot;</span>, <span class="at">hjust =</span> <span class="sc">-</span><span class="fl">0.5</span>, <span class="at">color =</span> <span class="st">&quot;blue&quot;</span>) <span class="sc">+</span> </span>
<span id="cb9-24"><a href="#cb9-24" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Elbow Plot Metrics&quot;</span>) <span class="sc">+</span></span>
<span id="cb9-25"><a href="#cb9-25" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">face =</span> <span class="st">&quot;bold&quot;</span>))</span></code></pre></div>
<p><img src="figure/6.1_Log_dimensional_reduction.Rmd/ElbowMetrics-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>In this case, the elbow might be observed around 15-25 PCs and the
calculated metrics are found in PCs 43 or 15 PCs. In the HBC training,
they suggest selecting the lower those metrics. However, it is usually
preferable to err on the higher side, so for this example the number of
selected PCs will be 30, as an intermediate between both (although it is
recommended to test downstream analysis using different numbers of
PCs).</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="do">## Code adapted from CosMxLite vignette</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="co"># Elbow plot</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="fu">ElbowPlot</span>(seu, <span class="at">ndims =</span> <span class="dv">50</span>) <span class="sc">+</span> </span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> npcs, <span class="at">linetype =</span> <span class="st">&quot;dotted&quot;</span>, <span class="at">color =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span> </span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> seu<span class="sc">@</span>reductions<span class="sc">$</span>pca<span class="sc">@</span>stdev[npcs], <span class="at">linetype =</span> <span class="st">&quot;dotted&quot;</span>, <span class="at">color =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Elbow Plot&quot;</span>) <span class="sc">+</span></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">face =</span> <span class="st">&quot;bold&quot;</span>))</span></code></pre></div>
<p><img src="figure/6.1_Log_dimensional_reduction.Rmd/ElbowPCA-1.png" width="864" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="umap" class="section level1">
<h1>UMAP</h1>
<p>In contrast to PCA, where there are up to 50 PCs generated through
the algorithm, UMAP is a non-linear dimensional reduction method that
generates only two coordinates which can be visualize in a 2D plot. In
standard single-cell expression analysis, UMAP is usually calculated
from the selected PCs previously calculated.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="do">## Code adapted from CosMxLite vignette</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="co"># Run UMAP</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>seu <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(seu, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span>npcs)</span></code></pre></div>
<div id="umap-dimplot" class="section level2">
<h2>UMAP DimPlot</h2>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="do">## Code adapted from CosMxLite vignette</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="co"># UMAP DimPlot</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="fu">DimPlot</span>(seu, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="at">cols =</span> <span class="st">&quot;stepped&quot;</span>, <span class="at">raster =</span> <span class="cn">FALSE</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code></pre></div>
<p><img src="figure/6.1_Log_dimensional_reduction.Rmd/VizUMAP1-1.png" width="384" style="display: block; margin: auto;" />
Currently, no clusters have been calculated, so every cell is
represented as part of one same group. However, at this point, UMAP
plotting can still be useful to visualize how are cells distributed
based on different variables, such as tissue (if there is more than one
tissue on the slide), slide ID (if there is more than one slide), total
counts or stains of interest.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="co"># Code inspired by Scratch Space vignette</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="co"># Random point order so no tissue or slide is on top by default </span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>random <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">Cells</span>(seu))</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a></span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a><span class="co"># By tissue (shown as an example, not applicable to this dataset)</span></span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a><span class="co"># p1 &lt;- DimPlot(seu, reduction = &quot;umap&quot;, cells = random, cols = &quot;stepped&quot;,</span></span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a><span class="co">#               raster = FALSE, group.by = &quot;tissue&quot;) +</span></span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a><span class="co">#   ggtitle(&quot;By tissue&quot;)</span></span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a></span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a><span class="co"># By slide ID (shown as an example, not applicable to this dataset)</span></span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a><span class="co"># p2 &lt;- DimPlot(seu, reduction = &quot;umap&quot;, cells = random, cols = &quot;stepped&quot;,</span></span>
<span id="cb13-13"><a href="#cb13-13" tabindex="-1"></a><span class="co">#               raster = FALSE, group.by = &quot;slide_ID_numeric&quot;) +</span></span>
<span id="cb13-14"><a href="#cb13-14" tabindex="-1"></a><span class="co">#   ggtitle(&quot;By Slide ID&quot;)</span></span>
<span id="cb13-15"><a href="#cb13-15" tabindex="-1"></a></span>
<span id="cb13-16"><a href="#cb13-16" tabindex="-1"></a><span class="do">## By total counts</span></span>
<span id="cb13-17"><a href="#cb13-17" tabindex="-1"></a>seu<span class="sc">$</span>log_nCount_RNA <span class="ot">&lt;-</span> <span class="fu">log2</span>(<span class="dv">1</span> <span class="sc">+</span> seu<span class="sc">$</span>nCount_RNA)</span>
<span id="cb13-18"><a href="#cb13-18" tabindex="-1"></a></span>
<span id="cb13-19"><a href="#cb13-19" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">FeaturePlot</span>(seu, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="at">cells =</span> random, </span>
<span id="cb13-20"><a href="#cb13-20" tabindex="-1"></a>                  <span class="at">features =</span> <span class="st">&quot;log_nCount_RNA&quot;</span>) <span class="sc">+</span></span>
<span id="cb13-21"><a href="#cb13-21" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>(<span class="at">option =</span> <span class="st">&quot;B&quot;</span>) <span class="sc">+</span> </span>
<span id="cb13-22"><a href="#cb13-22" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;By Total Counts&quot;</span>)</span>
<span id="cb13-23"><a href="#cb13-23" tabindex="-1"></a></span>
<span id="cb13-24"><a href="#cb13-24" tabindex="-1"></a><span class="do">## By DAPI stain</span></span>
<span id="cb13-25"><a href="#cb13-25" tabindex="-1"></a>seu<span class="sc">$</span>log_DAPI <span class="ot">&lt;-</span> <span class="fu">log2</span>(<span class="dv">1</span> <span class="sc">+</span> seu<span class="sc">$</span>Mean.DAPI)</span>
<span id="cb13-26"><a href="#cb13-26" tabindex="-1"></a></span>
<span id="cb13-27"><a href="#cb13-27" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> <span class="fu">FeaturePlot</span>(seu, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="at">cells =</span> random, </span>
<span id="cb13-28"><a href="#cb13-28" tabindex="-1"></a>                  <span class="at">features =</span> <span class="st">&quot;log_DAPI&quot;</span>) <span class="sc">+</span></span>
<span id="cb13-29"><a href="#cb13-29" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>(<span class="at">option =</span> <span class="st">&quot;B&quot;</span>) <span class="sc">+</span> </span>
<span id="cb13-30"><a href="#cb13-30" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;By DAPI stain&quot;</span>)</span>
<span id="cb13-31"><a href="#cb13-31" tabindex="-1"></a></span>
<span id="cb13-32"><a href="#cb13-32" tabindex="-1"></a><span class="co"># Arrange plots</span></span>
<span id="cb13-33"><a href="#cb13-33" tabindex="-1"></a>p3 <span class="sc">|</span> p4</span></code></pre></div>
<p><img src="figure/6.1_Log_dimensional_reduction.Rmd/VizUMAP2-1.png" width="1440" style="display: block; margin: auto;" /></p>
<p>This type is visualizations would help to determine if any technical
factor is influencing the data, i.e. batch effect related to the cells
being in different slides. For example, in this case, it can be seen
that all the cells with lower total counts are grouped in the middle of
the UMAP, which could indicate that further filtering would be
needed.</p>
</div>
<div id="save-the-seurat-object" class="section level2">
<h2>Save the Seurat object</h2>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co"># Save the Seurat object</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>name <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;seu_&quot;</span>, folder, <span class="st">&quot;_um.RDS&quot;</span>)</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="fu">saveRDS</span>(seu, <span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;processed_data&quot;</span>,folder,name))</span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="co"># Save the number of PCs used</span></span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a>name <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;npcs_&quot;</span>, folder, <span class="st">&quot;.RDS&quot;</span>)</span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a><span class="fu">saveRDS</span>(npcs, <span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;processed_data&quot;</span>,folder,name))</span></code></pre></div>
</div>
</div>
<div id="performance-and-session-info" class="section level1">
<h1>Performance and Session Info</h1>
<details>
<summary>
<strong>Performance Report</strong>
</summary>
<table class="table table-striped table-hover table-condensed" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
Chunk
</th>
<th style="text-align:right;">
Time_sec
</th>
<th style="text-align:right;">
Memory_Mb
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Libraries
</td>
<td style="text-align:right;">
1.46
</td>
<td style="text-align:right;">
148.8
</td>
</tr>
<tr>
<td style="text-align:left;">
LoadData
</td>
<td style="text-align:right;">
2.73
</td>
<td style="text-align:right;">
772.7
</td>
</tr>
<tr>
<td style="text-align:left;">
PCA
</td>
<td style="text-align:right;">
15.77
</td>
<td style="text-align:right;">
20.1
</td>
</tr>
<tr>
<td style="text-align:left;">
VizPCA1
</td>
<td style="text-align:right;">
0.48
</td>
<td style="text-align:right;">
11.5
</td>
</tr>
<tr>
<td style="text-align:left;">
VizPCA2
</td>
<td style="text-align:right;">
2.64
</td>
<td style="text-align:right;">
8.0
</td>
</tr>
<tr>
<td style="text-align:left;">
VizPCA3
</td>
<td style="text-align:right;">
1.77
</td>
<td style="text-align:right;">
6.3
</td>
</tr>
<tr>
<td style="text-align:left;">
ElbowMetrics
</td>
<td style="text-align:right;">
0.35
</td>
<td style="text-align:right;">
1.0
</td>
</tr>
<tr>
<td style="text-align:left;">
ElbowPCA
</td>
<td style="text-align:right;">
0.32
</td>
<td style="text-align:right;">
0.7
</td>
</tr>
<tr>
<td style="text-align:left;">
UMAP
</td>
<td style="text-align:right;">
66.76
</td>
<td style="text-align:right;">
3.5
</td>
</tr>
<tr>
<td style="text-align:left;">
VizUMAP1
</td>
<td style="text-align:right;">
1.34
</td>
<td style="text-align:right;">
7.3
</td>
</tr>
<tr>
<td style="text-align:left;">
VizUMAP2
</td>
<td style="text-align:right;">
2.84
</td>
<td style="text-align:right;">
17.4
</td>
</tr>
<tr>
<td style="text-align:left;">
SavingSeuObj
</td>
<td style="text-align:right;">
16.74
</td>
<td style="text-align:right;">
0.0
</td>
</tr>
<tr>
<td style="text-align:left;">
Total
</td>
<td style="text-align:right;">
113.20
</td>
<td style="text-align:right;">
997.3
</td>
</tr>
</tbody>
</table>
</details>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>R version 4.4.3 (2025-02-28 ucrt)
Platform: x86_64-w64-mingw32/x64
Running under: Windows 10 x64 (build 19045)

Matrix products: default


locale:
[1] LC_COLLATE=Spanish_Spain.utf8  LC_CTYPE=Spanish_Spain.utf8   
[3] LC_MONETARY=Spanish_Spain.utf8 LC_NUMERIC=C                  
[5] LC_TIME=Spanish_Spain.utf8    

time zone: Europe/Madrid
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] viridis_0.6.5      viridisLite_0.4.2  patchwork_1.3.0    ggplot2_3.5.1     
 [5] SeuratObject_4.1.4 Seurat_4.4.0       kableExtra_1.4.0   dplyr_1.1.4       
 [9] here_1.0.1         data.table_1.17.0  workflowr_1.7.1   

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3     rstudioapi_0.17.1      jsonlite_1.8.9        
  [4] magrittr_2.0.3         spatstat.utils_3.1-5   farver_2.1.2          
  [7] rmarkdown_2.29         fs_1.6.5               vctrs_0.6.5           
 [10] ROCR_1.0-11            spatstat.explore_3.5-2 htmltools_0.5.8.1     
 [13] sass_0.4.9             sctransform_0.4.2      parallelly_1.41.0     
 [16] KernSmooth_2.23-26     bslib_0.8.0            htmlwidgets_1.6.4     
 [19] ica_1.0-3              plyr_1.8.9             plotly_4.11.0         
 [22] zoo_1.8-14             cachem_1.1.0           whisker_0.4.1         
 [25] igraph_2.1.2           mime_0.12              lifecycle_1.0.4       
 [28] pkgconfig_2.0.3        Matrix_1.7-2           R6_2.5.1              
 [31] fastmap_1.2.0          fitdistrplus_1.2-4     future_1.34.0         
 [34] shiny_1.10.0           digest_0.6.37          colorspace_2.1-1      
 [37] ps_1.8.1               rprojroot_2.0.4        tensor_1.5.1          
 [40] irlba_2.3.5.1          labeling_0.4.3         progressr_0.15.1      
 [43] spatstat.sparse_3.1-0  httr_1.4.7             polyclip_1.10-7       
 [46] abind_1.4-8            compiler_4.4.3         withr_3.0.2           
 [49] MASS_7.3-64            tools_4.4.3            lmtest_0.9-40         
 [52] httpuv_1.6.15          future.apply_1.11.3    goftest_1.2-3         
 [55] glue_1.8.0             callr_3.7.6            nlme_3.1-167          
 [58] promises_1.3.2         grid_4.4.3             Rtsne_0.17            
 [61] getPass_0.2-4          cluster_2.1.8          reshape2_1.4.4        
 [64] generics_0.1.3         gtable_0.3.6           spatstat.data_3.1-6   
 [67] tidyr_1.3.1            sp_2.2-0               xml2_1.3.6            
 [70] spatstat.geom_3.5-0    RcppAnnoy_0.0.22       ggrepel_0.9.6         
 [73] RANN_2.6.2             pillar_1.10.0          stringr_1.5.1         
 [76] later_1.4.1            splines_4.4.3          lattice_0.22-6        
 [79] survival_3.8-3         deldir_2.0-4           tidyselect_1.2.1      
 [82] miniUI_0.1.1.1         pbapply_1.7-4          knitr_1.49            
 [85] git2r_0.36.2           gridExtra_2.3          svglite_2.1.3         
 [88] scattermore_1.2        xfun_0.49              matrixStats_1.4.1     
 [91] stringi_1.8.4          lazyeval_0.2.2         yaml_2.3.10           
 [94] evaluate_1.0.1         codetools_0.2-20       tibble_3.2.1          
 [97] cli_3.6.3              uwot_0.2.3             xtable_1.8-4          
[100] reticulate_1.40.0      systemfonts_1.1.0      munsell_0.5.1         
[103] processx_3.8.4         jquerylib_0.1.4        Rcpp_1.0.13-1         
[106] globals_0.16.3         spatstat.random_3.4-1  png_0.1-8             
[109] spatstat.univar_3.1-4  parallel_4.4.3         listenv_0.9.1         
[112] scales_1.3.0           ggridges_0.5.6         leiden_0.4.3.1        
[115] purrr_1.0.2            rlang_1.1.4            cowplot_1.1.3         </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
