<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>CosMx™ SMI Data Analysis Pipeline</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #204a87; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #204a87; font-weight: bold; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #ce5c00; font-weight: bold; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>







<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">CosMx™ SMI Data Analysis Pipeline</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="dataset.html">Dataset</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pipeline
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="pipeline.html">Introduction</a>
    </li>
    <li>
      <a href="0.0_data_loading.html">Data Loading</a>
    </li>
    <li>
      <a href="1.0_qc_and_filtering.html">QC and filtering</a>
    </li>
    <li>
      <a href="2.0_normalization.html">Normalization</a>
    </li>
    <li>
      <a href="3.0_dimensional_reduction.html">Dimensional reduction</a>
    </li>
    <li>
      <a href="4.0_insitutype_cell_typing.html">Cell typing</a>
    </li>
  </ul>
</li>
<li>
  <a href="norm_examples.html">Normalization</a>
</li>
<li>
  <a href="clus_examples.html">Cell typing</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
<li>
  <a href="https://github.com/lidiaga/CosMx_pipeline_LGA">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">CosMx™ SMI Data Analysis Pipeline</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-09-02
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>CosMx_pipeline_LGA/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20250517code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20250517)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20250517code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20250517)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrong9bc387c">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong> 9bc387c
</a>
</p>
</div>
<div id="strongRepositoryversionstrong9bc387c"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version 9bc387c.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    NBClust-Plots/
    Ignored:    analysis/.Rhistory
    Ignored:    data/seuAtoMx/
    Ignored:    output/processed_data/Log/
    Ignored:    output/processed_data/RC/
    Ignored:    output/processed_data/SCT/
    Ignored:    output/processed_data/exprMat_unfiltered.RDS
    Ignored:    output/processed_data/fov_positions_unfiltered.RDS
    Ignored:    output/processed_data/metadata_unfiltered.RDS
    Ignored:    output/processed_data/negMat_unfiltered.RDS
    Ignored:    output/processed_data/seu_filtered.RDS
    Ignored:    output/processed_data/seu_semifiltered.RDS

Untracked files:
    Untracked:  output/performance_reports/

Unstaged changes:
    Modified:   CosMx_pipeline_LGA.Rproj
    Modified:   _workflowr.yml

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/pipeline.Rmd</code>) and HTML
(<code>docs/pipeline.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
1cf5794
</td>
<td>
lga-uoc
</td>
<td>
2025-09-02
</td>
<td>
Edit website files
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p><a
href="https://nanostring.com/products/cosmx-spatial-molecular-imager/single-cell-imaging-overview/">CosMx™
SMI technology</a> is an imaged-based spatial transcriptomics technology
based on ISH and fluorescence techniques, and, as such, an important
step of processing its data would be processing the histology and
fluorescence images to extract the expression matrix and the spatial
location coordinates of cells and transcripts. However, for this
project, it was decided to start the pipeline from the data obtained
after segmentation and the construction of expression and localization
matrices, since, this steps can be efficiently carried out in <a
href="https://nanostring.com/products/atomx-spatial-informatics-platform/atomx-sip-overview/">AtoMx™
SIP</a> (the data-storing platform of the SMI instrument), avoiding the
need to download large amounts of image files. Additionally, if
open-source alternatives are preferred, there are already pipelines
available that allow this phase to be performed separately [1,2].</p>
<p>For the construction and structuring of this pipeline, a depth review
of open-source tools, available automatic pipelines and exploration of
various vignettes and practical usage examples was performed.
Additionally, resources and recommendations from the manufacturer,
NanoString® [3,4], as well as relevant technical documentation [5,6],
were also reviewed.</p>
<p>Finally, special mention to the <em>Scratch Space Vignette: Basics of
CosMx Analysis in R</em> [7] and the <em>CosMxLite vignette</em> [8],
which where very helpful for developing this work and understanding the
basics of this type of data analysis process. However, all references to
these and other sources have been commented and linked on the code when
needed.</p>
</div>
<div id="pipeline" class="section level1">
<h1>Pipeline</h1>
<p>The proposed pipeline consists of five main phases: data loading, QC
and filtering, normalization, dimensional reduction and cell typing:</p>
<div style="text-align: center;">
<p><img src="images/Pipeline_eng.svg" width="500" /></p>
</div>
<div style="font-size: smaller; text-align: justify;">
<p><strong>Figure 1. Pipeline flowchart.</strong> Created using
Lucidchart software. <strong>Blue:</strong> potential input files to the
pipeline; <strong>Dark gray:</strong> main processes; <strong>Light
gray:</strong> subprocesses; <strong>Orange:</strong> output files, with
those involved in later stages of the workflow highlighted in dark;
<strong>Solid lines:</strong> gray, general pipeline flow, and blue,
file input flow; <strong>Dashed lines:</strong> alternative or optional
flows. Red represents processes that were not developed in this work due
to time constraints, but could be included in future versions or
pipeline improvements, such as the multi-sample integration phase or the
semi-supervised clustering alternative.</p>
</div>
<p>With the goal of creating a flexible pipeline, the documents have
been parameterized, as far as possible, to allow for a variety of
options without affecting their proper functioning. However, the Cell
typing phase is subdivided into three independent scripts, given the
complexity of executing the different alternatives that the pipeline
contemplates:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Two starting points</strong> are available:</p>
<ol style="list-style-type: decimal">
<li><p>If starting from the AtoMx™ SIP exported flat files (.csv): start
in 0.0_data_loading.Rmd file.</p></li>
<li><p>If starting from the AtoMx™ SIP exported Seurat object: start in
1.0_qc_and_filtering.Rmd file.</p></li>
</ol></li>
<li><p><strong>Three normalization methods</strong> are available: RC,
Log and SCT (see <a href="norm_examples.html">Normalization</a> section
for more information)</p></li>
<li><p><strong>Three cell typing methods</strong> are available:
supervised classification with InSituType, unsupervied clustering with
InSitutype or unsupervised clustering with Seurat (see <a
href="clus_examples.html">Cell typing</a> section for more
information).</p></li>
</ol>
<div id="example-vignettes" class="section level2">
<h2>Example vignettes</h2>
<ol start="0" style="list-style-type: decimal">
<li><a href="0.0_data_loading.html">Data Loading</a></li>
<li><a href="1.0_qc_and_filtering.html">QC and filtering</a></li>
<li><a href="2.0_normalization.html">Normalization</a></li>
<li><a href="3.0_dimensional_reduction.html">Dimensional
reduction</a></li>
<li><a href="4.0_insitutype_cell_typing.html">Cell typing</a></li>
</ol>
</div>
<div id="total-time-and-memory" class="section level2">
<h2>Total time and memory</h2>
<table class="table table-striped table-hover table-condensed" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
Phase
</th>
<th style="text-align:right;">
Total_time_sec
</th>
<th style="text-align:right;">
Total_memory_Mb
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
0.0_data_loading_PR.csv
</td>
<td style="text-align:right;">
44.07
</td>
<td style="text-align:right;">
598.5
</td>
</tr>
<tr>
<td style="text-align:left;">
1.0_qc_and_filtering_PR.csv
</td>
<td style="text-align:right;">
67.28
</td>
<td style="text-align:right;">
905.7
</td>
</tr>
<tr>
<td style="text-align:left;">
2.0_normalization_PR.csv
</td>
<td style="text-align:right;">
112.90
</td>
<td style="text-align:right;">
1356.6
</td>
</tr>
<tr>
<td style="text-align:left;">
3.0_dimensional_reduction_PR.csv
</td>
<td style="text-align:right;">
194.35
</td>
<td style="text-align:right;">
1417.9
</td>
</tr>
<tr>
<td style="text-align:left;">
4.0_insitutype_cell_typing_PR.csv
</td>
<td style="text-align:right;">
727.21
</td>
<td style="text-align:right;">
2024.0
</td>
</tr>
<tr>
<td style="text-align:left;">
Total
</td>
<td style="text-align:right;">
1145.81
</td>
<td style="text-align:right;">
6302.7
</td>
</tr>
</tbody>
</table>
<p>In terms of time and memory usage, the example analysis shown in the
pipeline vignettes (using SCTransform as normalization method and
InSituType Supervised classification for cell typing) takes ~23 minutes
to execute, and 62,9 Gb of memory consumption.</p>
<p>However, other pipeline alternatives have also been executed in this
project and their total performances in time and memory are:</p>
<table class="table table-striped table-hover table-condensed" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
Alternative
</th>
<th style="text-align:right;">
Total_time_sec
</th>
<th style="text-align:right;">
Total_memory_Mb
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
RC
</td>
<td style="text-align:right;">
224.84
</td>
<td style="text-align:right;">
3436.2
</td>
</tr>
<tr>
<td style="text-align:left;">
Log
</td>
<td style="text-align:right;">
250.37
</td>
<td style="text-align:right;">
3436.2
</td>
</tr>
<tr>
<td style="text-align:left;">
SCT
</td>
<td style="text-align:right;">
418.60
</td>
<td style="text-align:right;">
4278.7
</td>
</tr>
<tr>
<td style="text-align:left;">
SCT_ISTSup
</td>
<td style="text-align:right;">
1145.81
</td>
<td style="text-align:right;">
6302.7
</td>
</tr>
<tr>
<td style="text-align:left;">
SCT_ISTUnsup
</td>
<td style="text-align:right;">
554.24
</td>
<td style="text-align:right;">
6284.9
</td>
</tr>
<tr>
<td style="text-align:left;">
SCT_SL
</td>
<td style="text-align:right;">
541.03
</td>
<td style="text-align:right;">
5875.0
</td>
</tr>
</tbody>
</table>
<p>The first three rows correspond to the total time an memory
consuption of the execution of the pipeline from data loading to
dimensional reduction, using each of the different normalization
alternatives:</p>
<ul>
<li><p><strong>RC:</strong> ~4 minutes and 34.2 Gb.</p></li>
<li><p><strong>Log:</strong> ~4 minutes and 34.2 Gb.</p></li>
<li><p><strong>SCT:</strong> ~5 minutes and 42.7 Gb.</p></li>
</ul>
<p>The last three rows correspond to the total time an memory consuption
of the execution of the pipeline from data loading to cell typing, using
the SCTransform normalization and each of the different cell typing
alternatives:</p>
<ul>
<li><p><strong>SCT + Supervised InSituType:</strong> ~23 minutes and
62,9 Gb.</p></li>
<li><p><strong>SCT + Unsupervised InSituType:</strong> ~10 minutes and
62,7 Gb.</p></li>
<li><p><strong>SCT + Seurat-Louvain:</strong> ~8 minutes and 58,6
Gb.</p></li>
</ul>
<hr />
</div>
</div>
<div id="bibliography" class="section level1">
<h1>Bibliography</h1>
<ol style="list-style-type: decimal">
<li><p>Blampey Q, Mulder K, Gardet M, Christodoulidis S, Dutertre CA,
André F, et al. Sopa: a technology-invariant pipeline for analyses of
image-based spatial omics. Nat Commun [Internet]. 2024 Jun 11 [cited
2025 Mar 3];15(1):4981. Available from: <a
href="https://www.nature.com/articles/s41467-024-48981-z"
class="uri">https://www.nature.com/articles/s41467-024-48981-z</a></p></li>
<li><p>Pollaris L, Vanneste B, Rombaut B, Defauw A, Vernaillen F,
Mortier J, et al. SPArrOW: a flexible, interactive and scalable pipeline
for spatial transcriptomics analysis [Internet]. 2024 [cited 2025 Mar
27]. Available from: <a
href="http://biorxiv.org/lookup/doi/10.1101/2024.07.04.601829"
class="uri">http://biorxiv.org/lookup/doi/10.1101/2024.07.04.601829</a></p></li>
<li><p>Bruker Spatial Biology, Inc. CosMx® Analysis Scratch Space
[Internet]. CosMx® Analysis Scratch Space. Available from: <a
href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/"
class="uri">https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/</a></p></li>
<li><p>GitHub - Nanostring-Biostats/CosMx-Analysis-Scratch-Space
[Internet]. CosMx-Analysis-Scratch-Space. Available from: <a
href="https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space"
class="uri">https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space</a></p></li>
<li><p>Bruker Spatial Biology. CosMx SMI Instrument User Manual
(Software v1.5) [Internet]. [cited 2025 Mar 21]. Available from: <a
href="https://university.nanostring.com/cosmx-smi-instrument-user-manual-software-v15"
class="uri">https://university.nanostring.com/cosmx-smi-instrument-user-manual-software-v15</a></p></li>
<li><p>Bruker Spatial Biology. CosMx SMI Data Analysis User Manual (on
AtoMx v1.4.0) [Internet]. [cited 2025 Mar 21]. Available from: <a
href="https://nanostring.com/resources/cosmx-smi-data-analysis-user-manual-software-v1-3/"
class="uri">https://nanostring.com/resources/cosmx-smi-data-analysis-user-manual-software-v1-3/</a></p></li>
<li><p>Danaher P. Vignette: Basics of CosMx Analysis in R – Blog
[Internet]. CosMx® Analysis Scratch Space. 2024 [cited 2025 Mar 27].
Available from: <a
href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/vignette-basic-analysis/"
class="uri">https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/vignette-basic-analysis/</a></p></li>
<li><p>O Hora B, Laddach R, Nuamah R, Chiappini C, Grigoriadis A, Quist
J. GitHub - cancerbioinformatics/CosMx_Lite [Internet]. CosMxLite. 2024
[cited 2025 Mar 27]. Available from: <a
href="https://github.com/cancerbioinformatics/CosMx_Lite"
class="uri">https://github.com/cancerbioinformatics/CosMx_Lite</a></p></li>
</ol>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>R version 4.4.3 (2025-02-28 ucrt)
Platform: x86_64-w64-mingw32/x64
Running under: Windows 10 x64 (build 19045)

Matrix products: default


locale:
[1] LC_COLLATE=Spanish_Spain.utf8  LC_CTYPE=Spanish_Spain.utf8   
[3] LC_MONETARY=Spanish_Spain.utf8 LC_NUMERIC=C                  
[5] LC_TIME=Spanish_Spain.utf8    

time zone: Europe/Madrid
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] kableExtra_1.4.0  dplyr_1.1.4       here_1.0.1        data.table_1.17.0
[5] workflowr_1.7.1  

loaded via a namespace (and not attached):
 [1] jsonlite_1.8.9    compiler_4.4.3    promises_1.3.2    tidyselect_1.2.1 
 [5] Rcpp_1.0.13-1     xml2_1.3.6        stringr_1.5.1     git2r_0.36.2     
 [9] callr_3.7.6       later_1.4.1       jquerylib_0.1.4   systemfonts_1.1.0
[13] scales_1.3.0      yaml_2.3.10       fastmap_1.2.0     R6_2.5.1         
[17] generics_0.1.3    knitr_1.49        tibble_3.2.1      munsell_0.5.1    
[21] rprojroot_2.0.4   svglite_2.1.3     bslib_0.8.0       pillar_1.10.0    
[25] rlang_1.1.4       cachem_1.1.0      stringi_1.8.4     httpuv_1.6.15    
[29] xfun_0.49         getPass_0.2-4     fs_1.6.5          sass_0.4.9       
[33] viridisLite_0.4.2 cli_3.6.3         magrittr_2.0.3    ps_1.8.1         
[37] digest_0.6.37     processx_3.8.4    rstudioapi_0.17.1 lifecycle_1.0.4  
[41] vctrs_0.6.5       evaluate_1.0.1    glue_1.8.0        whisker_0.4.1    
[45] colorspace_2.1-1  rmarkdown_2.29    httr_1.4.7        tools_4.4.3      
[49] pkgconfig_2.0.3   htmltools_0.5.8.1</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
